{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-09-02 11:55:23--  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.12, 130.14.250.31, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 103478058 (99M) [application/x-gzip]\n",
      "Saving to: ‘clinvar.vcf.gz’\n",
      "\n",
      "clinvar.vcf.gz      100%[===================>]  98.68M  28.0MB/s    in 3.7s    \n",
      "\n",
      "2024-09-02 11:55:28 (26.5 MB/s) - ‘clinvar.vcf.gz’ saved [103478058/103478058]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.1\n",
      "##fileDate=2024-08-25\n",
      "##source=ClinVar\n",
      "##reference=GRCh38\n",
      "##ID=<Description=\"ClinVar Variation ID\">\n",
      "##INFO=<ID=AF_ESP,Number=1,Type=Float,Description=\"allele frequencies from GO-ESP\">\n",
      "##INFO=<ID=AF_EXAC,Number=1,Type=Float,Description=\"allele frequencies from ExAC\">\n",
      "##INFO=<ID=AF_TGP,Number=1,Type=Float,Description=\"allele frequencies from TGP\">\n",
      "##INFO=<ID=ALLELEID,Number=1,Type=Integer,Description=\"the ClinVar Allele ID\">\n",
      "##INFO=<ID=CLNDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
      "##INFO=<ID=CLNDNINCL,Number=.,Type=String,Description=\"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
      "##INFO=<ID=CLNDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for germline classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=CLNDISDBINCL,Number=.,Type=String,Description=\"For included Variant: Tag-value pairs of disease database name and identifier for germline classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=CLNHGVS,Number=.,Type=String,Description=\"Top-level (primary assembly, alt, or patch) HGVS expression.\">\n",
      "##INFO=<ID=CLNREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of germline classification for the Variation ID\">\n",
      "##INFO=<ID=CLNSIG,Number=.,Type=String,Description=\"Aggregate germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNSIGCONF,Number=.,Type=String,Description=\"Conflicting germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNSIGINCL,Number=.,Type=String,Description=\"Germline classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNVC,Number=1,Type=String,Description=\"Variant type\">\n",
      "##INFO=<ID=CLNVCSO,Number=1,Type=String,Description=\"Sequence Ontology id for variant type\">\n",
      "##INFO=<ID=CLNVI,Number=.,Type=String,Description=\"the variant's clinical sources reported as tag-value pairs of database and variant identifier\">\n",
      "##INFO=<ID=DBVARID,Number=.,Type=String,Description=\"nsv accessions from dbVar for the variant\">\n",
      "##INFO=<ID=GENEINFO,Number=1,Type=String,Description=\"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\">\n",
      "##INFO=<ID=MC,Number=.,Type=String,Description=\"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\">\n",
      "##INFO=<ID=ONCDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDB\">\n",
      "##INFO=<ID=ONCDNINCL,Number=.,Type=String,Description=\"For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDBINCL\">\n",
      "##INFO=<ID=ONCDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for oncogenicity classifications, e.g. MedGen:NNNNNN\">\n",
      "##INFO=<ID=ONCDISDBINCL,Number=.,Type=String,Description=\"For included variant: Tag-value pairs of disease database name and identifier for oncogenicity classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=ONC,Number=.,Type=String,Description=\"Aggregate oncogenicity classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=ONCINCL,Number=.,Type=String,Description=\"Oncogenicity classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat clinvar.vcf.gz | head -n 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'Y' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'MT' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113889.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187633.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187661.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187693.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_009646201.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 16\u001b[0m\n\u001b[1;32m     13\u001b[0m     review_status \u001b[38;5;241m=\u001b[39m variant\u001b[38;5;241m.\u001b[39mINFO\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCLNREVSTAT\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     14\u001b[0m     rows\u001b[38;5;241m.\u001b[39mappend([variant\u001b[38;5;241m.\u001b[39mCHROM, variant\u001b[38;5;241m.\u001b[39mPOS, variant\u001b[38;5;241m.\u001b[39mREF, variant\u001b[38;5;241m.\u001b[39mALT[\u001b[38;5;241m0\u001b[39m], cln_sig, variant\u001b[38;5;241m.\u001b[39mID, review_status, consequence])\n\u001b[0;32m---> 16\u001b[0m V \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(rows, columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchrom\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpos\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mref\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124malt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlabel\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mid\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreview_status\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconsequence\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m     17\u001b[0m V \u001b[38;5;241m=\u001b[39m V[V\u001b[38;5;241m.\u001b[39mchrom\u001b[38;5;241m!=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMT\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m     18\u001b[0m V \u001b[38;5;241m=\u001b[39m V[V\u001b[38;5;241m.\u001b[39malt \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mN\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "from cyvcf2 import VCF\n",
    "\n",
    "rows = []\n",
    "for variant in VCF(\"clinvar.vcf.gz\"):\n",
    "    if variant.INFO.get(\"CLNVC\") != \"single_nucleotide_variant\": continue\n",
    "    if len(variant.ALT) != 1: continue\n",
    "    cln_sig = variant.INFO.get(\"CLNSIG\")\n",
    "    if cln_sig not in [\"Pathogenic\", \"Likely_Pathogenic\"]: continue\n",
    "    MC = variant.INFO.get(\"MC\")\n",
    "    if MC is None: continue\n",
    "    consequences = [x.split(\"|\")[1] for x in MC.split(\",\")]\n",
    "    consequence = ','.join(np.unique(consequences))\n",
    "    review_status = variant.INFO.get(\"CLNREVSTAT\")\n",
    "    rows.append([variant.CHROM, variant.POS, variant.REF, variant.ALT[0], cln_sig, variant.ID, review_status, consequence])\n",
    "\n",
    "V = pd.DataFrame(rows, columns=[\"chrom\", \"pos\", \"ref\", \"alt\", \"label\", \"id\", \"review_status\", \"consequence\"])\n",
    "V = V[V.chrom!=\"MT\"]\n",
    "V = V[V.alt != \"N\"]\n",
    "V"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>label</th>\n",
       "      <th>id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>consequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>943995</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>950448</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>976215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1320032</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>183381</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1022368</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>3255595</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>126556</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75743</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787515</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>492908</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75744</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787551</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9754</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75745</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787592</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9751</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75746</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787600</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>9753</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>nonsense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75747</th>\n",
       "      <td>Y</td>\n",
       "      <td>7063898</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>625467</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75733 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      pos ref alt       label       id  \\\n",
       "0         1   943995   C   T  Pathogenic   950448   \n",
       "1         1   976215   A   G  Pathogenic  1320032   \n",
       "2         1  1014143   C   T  Pathogenic   183381   \n",
       "3         1  1022368   C   A  Pathogenic  3255595   \n",
       "4         1  1041582   C   T  Pathogenic   126556   \n",
       "...     ...      ...  ..  ..         ...      ...   \n",
       "75743     Y  2787515   C   A  Pathogenic   492908   \n",
       "75744     Y  2787551   C   T  Pathogenic     9754   \n",
       "75745     Y  2787592   A   T  Pathogenic     9751   \n",
       "75746     Y  2787600   G   A  Pathogenic     9753   \n",
       "75747     Y  7063898   A   T  Pathogenic   625467   \n",
       "\n",
       "                             review_status       consequence  \n",
       "0      criteria_provided,_single_submitter          nonsense  \n",
       "1           no_assertion_criteria_provided  missense_variant  \n",
       "2           no_assertion_criteria_provided          nonsense  \n",
       "3      criteria_provided,_single_submitter          nonsense  \n",
       "4           no_assertion_criteria_provided          nonsense  \n",
       "...                                    ...               ...  \n",
       "75743       no_assertion_criteria_provided  missense_variant  \n",
       "75744       no_assertion_criteria_provided  missense_variant  \n",
       "75745       no_assertion_criteria_provided          nonsense  \n",
       "75746       no_assertion_criteria_provided          nonsense  \n",
       "75747       no_assertion_criteria_provided  missense_variant  \n",
       "\n",
       "[75733 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>label</th>\n",
       "      <th>id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>consequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>1043197</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1284256</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1</td>\n",
       "      <td>1787438</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1315579</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1</td>\n",
       "      <td>1789052</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>2703061</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1</td>\n",
       "      <td>1804582</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1315576</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>1</td>\n",
       "      <td>2408451</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>6770</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75683</th>\n",
       "      <td>X</td>\n",
       "      <td>154984804</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1698780</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75684</th>\n",
       "      <td>X</td>\n",
       "      <td>154984805</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>10182</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75687</th>\n",
       "      <td>X</td>\n",
       "      <td>154987307</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>10162</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75697</th>\n",
       "      <td>X</td>\n",
       "      <td>154996972</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1330755</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75717</th>\n",
       "      <td>X</td>\n",
       "      <td>155264073</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>10542</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11727 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom        pos ref alt       label       id  \\\n",
       "8         1    1043197   G   C  Pathogenic  1284256   \n",
       "40        1    1787438   C   A  Pathogenic  1315579   \n",
       "41        1    1789052   C   A  Pathogenic  2703061   \n",
       "48        1    1804582   C   A  Pathogenic  1315576   \n",
       "81        1    2408451   C   T  Pathogenic     6770   \n",
       "...     ...        ...  ..  ..         ...      ...   \n",
       "75683     X  154984804   C   T  Pathogenic  1698780   \n",
       "75684     X  154984805   T   C  Pathogenic    10182   \n",
       "75687     X  154987307   T   C  Pathogenic    10162   \n",
       "75697     X  154996972   C   T  Pathogenic  1330755   \n",
       "75717     X  155264073   C   T  Pathogenic    10542   \n",
       "\n",
       "                                           review_status  \\\n",
       "8                         no_assertion_criteria_provided   \n",
       "40                        no_assertion_criteria_provided   \n",
       "41                   criteria_provided,_single_submitter   \n",
       "48                        no_assertion_criteria_provided   \n",
       "81     criteria_provided,_multiple_submitters,_no_con...   \n",
       "...                                                  ...   \n",
       "75683                criteria_provided,_single_submitter   \n",
       "75684                     no_assertion_criteria_provided   \n",
       "75687                     no_assertion_criteria_provided   \n",
       "75697                criteria_provided,_single_submitter   \n",
       "75717                     no_assertion_criteria_provided   \n",
       "\n",
       "                   consequence  \n",
       "8               intron_variant  \n",
       "40     splice_acceptor_variant  \n",
       "41        splice_donor_variant  \n",
       "48     splice_acceptor_variant  \n",
       "81        splice_donor_variant  \n",
       "...                        ...  \n",
       "75683  splice_acceptor_variant  \n",
       "75684  splice_acceptor_variant  \n",
       "75687  splice_acceptor_variant  \n",
       "75697     splice_donor_variant  \n",
       "75717     splice_donor_variant  \n",
       "\n",
       "[11727 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2 = V[\n",
    "    ~(V.consequence.str.contains(\"missense\")) &\n",
    "    ~(V.consequence.str.contains(\"coding\")) &\n",
    "    ~(V.consequence.str.contains(\"nonsense\")) &\n",
    "    ~(V.consequence.str.contains(\"UTR\"))\n",
    "]\n",
    "V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "consequence\n",
       "splice_donor_variant                                                                 5223\n",
       "splice_acceptor_variant                                                              3961\n",
       "intron_variant                                                                       1294\n",
       "intron_variant,splice_donor_variant                                                   505\n",
       "intron_variant,splice_acceptor_variant                                                419\n",
       "synonymous_variant                                                                    225\n",
       "stop_lost                                                                              49\n",
       "intron_variant,synonymous_variant                                                      24\n",
       "splice_donor_variant,synonymous_variant                                                 7\n",
       "splice_acceptor_variant,synonymous_variant                                              5\n",
       "intron_variant,stop_lost                                                                4\n",
       "intron_variant,splice_donor_variant,synonymous_variant                                  3\n",
       "genic_upstream_transcript_variant                                                       3\n",
       "genic_upstream_transcript_variant,initiator_codon_variant,splice_acceptor_variant       1\n",
       "genic_upstream_transcript_variant,intron_variant                                        1\n",
       "splice_donor_variant,stop_lost                                                          1\n",
       "initiator_codon_variant,intron_variant,splice_donor_variant                             1\n",
       "genic_downstream_transcript_variant,intron_variant                                      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2.consequence.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gpn",
   "language": "python",
   "name": "gpn"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
