{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c3d39fe4-48bb-4c83-afe7-3e03edf37139",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a47299cb-af82-407b-b44a-ddebc1c2a278",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!wget https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/submission_summary.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b09a40a5-46fa-412c-a320-b922d509391d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/users/gbenegas/software/mambaforge/envs/gpn/lib/python3.11/pty.py:89: RuntimeWarning: Using fork() can cause Polars to deadlock in the child process.\n",
      "In addition, using fork() with Python in general is a recipe for mysterious\n",
      "deadlocks and crashes.\n",
      "\n",
      "The most likely reason you are seeing this error is because you are using the\n",
      "multiprocessing module on Linux, which uses fork() by default. This will be\n",
      "fixed in Python 3.14. Until then, you want to use the \"spawn\" context instead.\n",
      "\n",
      "See https://docs.pola.rs/user-guide/misc/multiprocessing/ for details.\n",
      "\n",
      "If you really know what your doing, you can silence this warning with the warning module\n",
      "or by setting POLARS_ALLOW_FORKING_THREAD=1.\n",
      "\n",
      "  pid, fd = os.forkpty()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##Overview of interpretation, phenotypes, observations, and methods reported in each current submission \n",
      "##Explanation of the columns in this report\n",
      "#VariationID:                   the identifier assigned by ClinVar and used to build the URL, namely https://ncbi.nlm.nih.gov/clinvar/VariationID\n",
      "#ClinicalSignificance:          the germline classification on this submitted record\n",
      "#DateLastEvaluated:             the last date the classification on this record was evaluated by the submitter\n",
      "#Description:                   an optional free text description comment describing the rationale for the classification\n",
      "#SubmittedPhenotypeInfo:        the name(s) or identifier(s) submitted as the condition for the classification \n",
      "#ReportedPhenotypeInfo:         the MedGen identifier/name combinations that the submitted condition for the classification maps to. 'na' means there is no public identifer in MedGen for the condition.\n",
      "#ReviewStatus:                  the level of review for this submitted record; see http//www.ncbi.nlm.nih.gov/clinvar/docs/variation_report/#review_status\n",
      "#CollectionMethod:              the method by which the submitter collected the data for the classification; see https://www.ncbi.nlm.nih.gov/clinvar/docs/spreadsheet/#collection\n",
      "#OriginCounts:                  the allele origin reported by the submitter and the number of observations for each origin. ‘na’ means that the number of observations was not provided by the submitter.\n",
      "#Submitter:                     the submitter of this record\n",
      "#SCV:                            the accession and current version assigned by ClinVar to this submitted record\n",
      "#SubmittedGeneSymbol:           the gene symbol reported in this submitted record\n",
      "#ExplanationOfInterpretation:   more details if the germline classification (ClinicalSignificance) is 'other' or 'drug response'\n",
      "#SomaticClinicalImpact:         the somatic classification of clinical impact on this submitted record\n",
      "#Oncogenicity:                  the somatic classification of oncogenicity on this submitted record\n",
      "#ContributesToAggregateClassification: indicate if the SCV contributes to the aggregate classification\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat submission_summary.txt.gz | head -n 18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "53b97529-4ab4-485b-828e-ab57e02df839",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5_399_159, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>SubmittedPhenotypeInfo</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;OMIM:613647&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;clinical testing&quot;</td><td>&quot;Paris Brain Institute, Inserm …</td></tr><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;not provided&quot;</td><td>&quot;C3661900:not provided&quot;</td><td>&quot;clinical testing&quot;</td><td>&quot;Athena Diagnostics&quot;</td></tr><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48, AUTOSOM…</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>2</td><td>&quot;Likely pathogenic&quot;</td><td>&quot;Macular dystrophy with or with…</td><td>&quot;na:Macular dystrophy with or w…</td><td>&quot;research&quot;</td><td>&quot;Ophthalmic Genetics Group, Ins…</td></tr><tr><td>3</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3900693</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900694</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900695</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900696</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900697</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5_399_159, 6)\n",
       "┌──────────────┬────────────────┬────────────────┬────────────────┬────────────────┬───────────────┐\n",
       "│ #VariationID ┆ ClinicalSignif ┆ SubmittedPheno ┆ ReportedPhenot ┆ CollectionMeth ┆ Submitter     │\n",
       "│ ---          ┆ icance         ┆ typeInfo       ┆ ypeInfo        ┆ od             ┆ ---           │\n",
       "│ i64          ┆ ---            ┆ ---            ┆ ---            ┆ ---            ┆ str           │\n",
       "│              ┆ str            ┆ str            ┆ str            ┆ str            ┆               │\n",
       "╞══════════════╪════════════════╪════════════════╪════════════════╪════════════════╪═══════════════╡\n",
       "│ 2            ┆ Pathogenic     ┆ OMIM:613647    ┆ C3150901:Hered ┆ clinical       ┆ Paris Brain   │\n",
       "│              ┆                ┆                ┆ itary spastic  ┆ testing        ┆ Institute,    │\n",
       "│              ┆                ┆                ┆ pa…            ┆                ┆ Inserm …      │\n",
       "│ 2            ┆ Pathogenic     ┆ not provided   ┆ C3661900:not   ┆ clinical       ┆ Athena        │\n",
       "│              ┆                ┆                ┆ provided       ┆ testing        ┆ Diagnostics   │\n",
       "│ 2            ┆ Pathogenic     ┆ SPASTIC        ┆ C3150901:Hered ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆ PARAPLEGIA 48, ┆ itary spastic  ┆ only           ┆               │\n",
       "│              ┆                ┆ AUTOSOM…       ┆ pa…            ┆                ┆               │\n",
       "│ 2            ┆ Likely         ┆ Macular        ┆ na:Macular     ┆ research       ┆ Ophthalmic    │\n",
       "│              ┆ pathogenic     ┆ dystrophy with ┆ dystrophy with ┆                ┆ Genetics      │\n",
       "│              ┆                ┆ or with…       ┆ or w…          ┆                ┆ Group, Ins…   │\n",
       "│ 3            ┆ Pathogenic     ┆ SPASTIC        ┆ C3150901:Hered ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆ PARAPLEGIA 48  ┆ itary spastic  ┆ only           ┆               │\n",
       "│              ┆                ┆                ┆ pa…            ┆                ┆               │\n",
       "│ …            ┆ …              ┆ …              ┆ …              ┆ …              ┆ …             │\n",
       "│ 3900693      ┆ Pathogenic     ┆ -              ┆ na:GUILLOUET-G ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆                ┆ ORDON SYNDROME ┆ only           ┆               │\n",
       "│ 3900694      ┆ Pathogenic     ┆ -              ┆ na:GUILLOUET-G ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆                ┆ ORDON SYNDROME ┆ only           ┆               │\n",
       "│ 3900695      ┆ Pathogenic     ┆ -              ┆ na:GUILLOUET-G ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆                ┆ ORDON SYNDROME ┆ only           ┆               │\n",
       "│ 3900696      ┆ Pathogenic     ┆ -              ┆ na:GUILLOUET-G ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆                ┆ ORDON SYNDROME ┆ only           ┆               │\n",
       "│ 3900697      ┆ Pathogenic     ┆ -              ┆ na:GUILLOUET-G ┆ literature     ┆ OMIM          │\n",
       "│              ┆                ┆                ┆ ORDON SYNDROME ┆ only           ┆               │\n",
       "└──────────────┴────────────────┴────────────────┴────────────────┴────────────────┴───────────────┘"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pl.read_csv(\n",
    "    \"submission_summary.txt.gz\", separator=\"\\t\", skip_rows=18,\n",
    "    columns=[\n",
    "        \"#VariationID\",\n",
    "        \"ClinicalSignificance\",\n",
    "        \"SubmittedPhenotypeInfo\",\n",
    "        \"ReportedPhenotypeInfo\",\n",
    "        \"CollectionMethod\",\n",
    "        \"Submitter\",\n",
    "    ],\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "79f0dd38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5458\tBenign\tFeb 22, 2021\tThis variant is classified as benign based on ACMG/AMP sequence variant interpretation guidelines (Richards et al. 2015 PMID: 25741868, with internal and published modifications).\tKALRN-related condition\tna:KALRN-related disorder\tno assertion criteria provided\tclinical testing\tgermline:na\tPreventionGenetics, part of Exact Sciences\tSCV004721556.2\tKALRN\t-\t-\t-\tno\n",
      "5458\tBenign\tFeb 07, 2025\tThis variant is considered likely benign or benign based on one or more of the following: it is predicted to be benign by multiple in silico algorithms, and/or has population frequency not consistent with disease, and/or has normal protein function, and/or has lack of segregation with disease, and/or has been detected in co-occurrence with known pathogenic variant, and/or has lack of disease association in case-control studies, and/or is located in a region inconsistent with a known cause of pathogenicity. GnomAD 4.1.0 frequency 0.1421 2094 homozygotes\tMedGen:C1837173\tC1837173:Coronary heart disease, susceptibility to, 5\tcriteria provided, single submitter\tclinical testing\tunknown:na\tMendelics\tSCV005870925.1\tKALRN\t-\t-\t-\tyes\n",
      "5458\tPathogenic\tApr 01, 2007\t-\tRECLASSIFIED - VARIANT of UNKNOWN SIGNIFICANCE\tC1837173:Coronary heart disease, susceptibility to, 5\tno assertion criteria provided\tliterature only\tgermline:na\tOMIM\tSCV000025973.3\tKALRN\t-\t-\t-\tno\n"
     ]
    }
   ],
   "source": [
    "# I guess need to pick up the column that says \"RECLASSIFIED\"\n",
    "!zcat submission_summary.txt.gz | awk '$1 == 5458'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "d9f77871-4b62-437d-ab43-cb43bf7b2964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (36_803, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>SubmittedPhenotypeInfo</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48, AUTOSOM…</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>4</td><td>&quot;Uncertain significance&quot;</td><td>&quot;RECLASSIFIED - VARIANT OF UNKN…</td><td>&quot;C4551772:Galloway-Mowat syndro…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>5</td><td>&quot;Pathogenic&quot;</td><td>&quot;MITOCHONDRIAL COMPLEX I DEFICI…</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>6</td><td>&quot;Pathogenic&quot;</td><td>&quot;MITOCHONDRIAL COMPLEX I DEFICI…</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3900693</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900694</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900695</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900696</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900697</td><td>&quot;Pathogenic&quot;</td><td>&quot;-&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (36_803, 6)\n",
       "┌──────────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignifi ┆ SubmittedPhenot ┆ ReportedPhenoty ┆ CollectionMetho ┆ Submitter │\n",
       "│ ---          ┆ cance           ┆ ypeInfo         ┆ peInfo          ┆ d               ┆ ---       │\n",
       "│ i64          ┆ ---             ┆ ---             ┆ ---             ┆ ---             ┆ str       │\n",
       "│              ┆ str             ┆ str             ┆ str             ┆ str             ┆           │\n",
       "╞══════════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪═══════════╡\n",
       "│ 2            ┆ Pathogenic      ┆ SPASTIC         ┆ C3150901:Heredi ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ PARAPLEGIA 48,  ┆ tary spastic    ┆                 ┆           │\n",
       "│              ┆                 ┆ AUTOSOM…        ┆ pa…             ┆                 ┆           │\n",
       "│ 3            ┆ Pathogenic      ┆ SPASTIC         ┆ C3150901:Heredi ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ PARAPLEGIA 48   ┆ tary spastic    ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ pa…             ┆                 ┆           │\n",
       "│ 4            ┆ Uncertain       ┆ RECLASSIFIED -  ┆ C4551772:Gallow ┆ literature only ┆ OMIM      │\n",
       "│              ┆ significance    ┆ VARIANT OF      ┆ ay-Mowat        ┆                 ┆           │\n",
       "│              ┆                 ┆ UNKN…           ┆ syndro…         ┆                 ┆           │\n",
       "│ 5            ┆ Pathogenic      ┆ MITOCHONDRIAL   ┆ C4748791:Mitoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ COMPLEX I       ┆ ondrial         ┆                 ┆           │\n",
       "│              ┆                 ┆ DEFICI…         ┆ complex…        ┆                 ┆           │\n",
       "│ 6            ┆ Pathogenic      ┆ MITOCHONDRIAL   ┆ C4748791:Mitoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ COMPLEX I       ┆ ondrial         ┆                 ┆           │\n",
       "│              ┆                 ┆ DEFICI…         ┆ complex…        ┆                 ┆           │\n",
       "│ …            ┆ …               ┆ …               ┆ …               ┆ …               ┆ …         │\n",
       "│ 3900693      ┆ Pathogenic      ┆ -               ┆ na:GUILLOUET-GO ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆                 ┆ RDON SYNDROME   ┆                 ┆           │\n",
       "│ 3900694      ┆ Pathogenic      ┆ -               ┆ na:GUILLOUET-GO ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆                 ┆ RDON SYNDROME   ┆                 ┆           │\n",
       "│ 3900695      ┆ Pathogenic      ┆ -               ┆ na:GUILLOUET-GO ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆                 ┆ RDON SYNDROME   ┆                 ┆           │\n",
       "│ 3900696      ┆ Pathogenic      ┆ -               ┆ na:GUILLOUET-GO ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆                 ┆ RDON SYNDROME   ┆                 ┆           │\n",
       "│ 3900697      ┆ Pathogenic      ┆ -               ┆ na:GUILLOUET-GO ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆                 ┆ RDON SYNDROME   ┆                 ┆           │\n",
       "└──────────────┴─────────────────┴─────────────────┴─────────────────┴─────────────────┴───────────┘"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.filter(Submitter=\"OMIM\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c7d3f706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>SubmittedPhenotypeInfo</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>5458</td><td>&quot;Pathogenic&quot;</td><td>&quot;RECLASSIFIED - VARIANT of UNKN…</td><td>&quot;C1837173:Coronary heart diseas…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 6)\n",
       "┌──────────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignifi ┆ SubmittedPhenot ┆ ReportedPhenoty ┆ CollectionMetho ┆ Submitter │\n",
       "│ ---          ┆ cance           ┆ ypeInfo         ┆ peInfo          ┆ d               ┆ ---       │\n",
       "│ i64          ┆ ---             ┆ ---             ┆ ---             ┆ ---             ┆ str       │\n",
       "│              ┆ str             ┆ str             ┆ str             ┆ str             ┆           │\n",
       "╞══════════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪═══════════╡\n",
       "│ 5458         ┆ Pathogenic      ┆ RECLASSIFIED -  ┆ C1837173:Corona ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ VARIANT of      ┆ ry heart        ┆                 ┆           │\n",
       "│              ┆                 ┆ UNKN…           ┆ diseas…         ┆                 ┆           │\n",
       "└──────────────┴─────────────────┴─────────────────┴─────────────────┴─────────────────┴───────────┘"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for some reason, this one passed the filter??\n",
    "# this one was reclassified\n",
    "# so need to find where to find the reclassification note from OMIM itself (not clinvar)\n",
    "df.filter(pl.col(\"#VariationID\") == 5458)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "3aad2c8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (20, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>SubmittedPhenotypeInfo</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48, AUTOSOM…</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3</td><td>&quot;Pathogenic&quot;</td><td>&quot;SPASTIC PARAPLEGIA 48&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>4</td><td>&quot;Uncertain significance&quot;</td><td>&quot;RECLASSIFIED - VARIANT OF UNKN…</td><td>&quot;C4551772:Galloway-Mowat syndro…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>5</td><td>&quot;Pathogenic&quot;</td><td>&quot;MITOCHONDRIAL COMPLEX I DEFICI…</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>6</td><td>&quot;Pathogenic&quot;</td><td>&quot;MITOCHONDRIAL COMPLEX I DEFICI…</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>18</td><td>&quot;Pathogenic&quot;</td><td>&quot;HEMOCHROMATOSIS, TYPE 1&quot;</td><td>&quot;C3469186:Hemochromatosis type …</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>19</td><td>&quot;Pathogenic&quot;</td><td>&quot;HEMOCHROMATOSIS, TYPE 1&quot;</td><td>&quot;C3469186:Hemochromatosis type …</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>20</td><td>&quot;Pathogenic&quot;</td><td>&quot;CRANIOECTODERMAL DYSPLASIA 2&quot;</td><td>&quot;C3150874:Cranioectodermal dysp…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>21</td><td>&quot;Pathogenic&quot;</td><td>&quot;CRANIOECTODERMAL DYSPLASIA 2&quot;</td><td>&quot;C3150874:Cranioectodermal dysp…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>22</td><td>&quot;Pathogenic&quot;</td><td>&quot;CRANIOECTODERMAL DYSPLASIA 2&quot;</td><td>&quot;C3150874:Cranioectodermal dysp…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (20, 6)\n",
       "┌──────────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignifi ┆ SubmittedPhenot ┆ ReportedPhenoty ┆ CollectionMetho ┆ Submitter │\n",
       "│ ---          ┆ cance           ┆ ypeInfo         ┆ peInfo          ┆ d               ┆ ---       │\n",
       "│ i64          ┆ ---             ┆ ---             ┆ ---             ┆ ---             ┆ str       │\n",
       "│              ┆ str             ┆ str             ┆ str             ┆ str             ┆           │\n",
       "╞══════════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪═══════════╡\n",
       "│ 2            ┆ Pathogenic      ┆ SPASTIC         ┆ C3150901:Heredi ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ PARAPLEGIA 48,  ┆ tary spastic    ┆                 ┆           │\n",
       "│              ┆                 ┆ AUTOSOM…        ┆ pa…             ┆                 ┆           │\n",
       "│ 3            ┆ Pathogenic      ┆ SPASTIC         ┆ C3150901:Heredi ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ PARAPLEGIA 48   ┆ tary spastic    ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ pa…             ┆                 ┆           │\n",
       "│ 4            ┆ Uncertain       ┆ RECLASSIFIED -  ┆ C4551772:Gallow ┆ literature only ┆ OMIM      │\n",
       "│              ┆ significance    ┆ VARIANT OF      ┆ ay-Mowat        ┆                 ┆           │\n",
       "│              ┆                 ┆ UNKN…           ┆ syndro…         ┆                 ┆           │\n",
       "│ 5            ┆ Pathogenic      ┆ MITOCHONDRIAL   ┆ C4748791:Mitoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ COMPLEX I       ┆ ondrial         ┆                 ┆           │\n",
       "│              ┆                 ┆ DEFICI…         ┆ complex…        ┆                 ┆           │\n",
       "│ 6            ┆ Pathogenic      ┆ MITOCHONDRIAL   ┆ C4748791:Mitoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ COMPLEX I       ┆ ondrial         ┆                 ┆           │\n",
       "│              ┆                 ┆ DEFICI…         ┆ complex…        ┆                 ┆           │\n",
       "│ …            ┆ …               ┆ …               ┆ …               ┆ …               ┆ …         │\n",
       "│ 18           ┆ Pathogenic      ┆ HEMOCHROMATOSIS ┆ C3469186:Hemoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ , TYPE 1        ┆ romatosis type  ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ …               ┆                 ┆           │\n",
       "│ 19           ┆ Pathogenic      ┆ HEMOCHROMATOSIS ┆ C3469186:Hemoch ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ , TYPE 1        ┆ romatosis type  ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ …               ┆                 ┆           │\n",
       "│ 20           ┆ Pathogenic      ┆ CRANIOECTODERMA ┆ C3150874:Cranio ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ L DYSPLASIA 2   ┆ ectodermal      ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ dysp…           ┆                 ┆           │\n",
       "│ 21           ┆ Pathogenic      ┆ CRANIOECTODERMA ┆ C3150874:Cranio ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ L DYSPLASIA 2   ┆ ectodermal      ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ dysp…           ┆                 ┆           │\n",
       "│ 22           ┆ Pathogenic      ┆ CRANIOECTODERMA ┆ C3150874:Cranio ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ L DYSPLASIA 2   ┆ ectodermal      ┆                 ┆           │\n",
       "│              ┆                 ┆                 ┆ dysp…           ┆                 ┆           │\n",
       "└──────────────┴─────────────────┴─────────────────┴─────────────────┴─────────────────┴───────────┘"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "52cc8c1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>SubmittedPhenotypeInfo</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>5458</td><td>&quot;Pathogenic&quot;</td><td>&quot;RECLASSIFIED - VARIANT of UNKN…</td><td>&quot;C1837173:Coronary heart diseas…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 6)\n",
       "┌──────────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignifi ┆ SubmittedPhenot ┆ ReportedPhenoty ┆ CollectionMetho ┆ Submitter │\n",
       "│ ---          ┆ cance           ┆ ypeInfo         ┆ peInfo          ┆ d               ┆ ---       │\n",
       "│ i64          ┆ ---             ┆ ---             ┆ ---             ┆ ---             ┆ str       │\n",
       "│              ┆ str             ┆ str             ┆ str             ┆ str             ┆           │\n",
       "╞══════════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪═══════════╡\n",
       "│ 5458         ┆ Pathogenic      ┆ RECLASSIFIED -  ┆ C1837173:Corona ┆ literature only ┆ OMIM      │\n",
       "│              ┆                 ┆ VARIANT of      ┆ ry heart        ┆                 ┆           │\n",
       "│              ┆                 ┆ UNKN…           ┆ diseas…         ┆                 ┆           │\n",
       "└──────────────┴─────────────────┴─────────────────┴─────────────────┴─────────────────┴───────────┘"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I guess it was only that one that was reclassified\n",
    "df.filter(pl.col(\"ClinicalSignificance\") == \"Pathogenic\", pl.col(\"SubmittedPhenotypeInfo\").str.contains(\"RECLASSIFIED\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ed8c7af5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['C1837173:Coronary heart disease, susceptibility to, 5'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"#VariationID\") == 5458)[\"ReportedPhenotypeInfo\"].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "dabcfc49-203f-4540-9cad-e96d3c1797cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (36_148, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3</td><td>&quot;Pathogenic&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>4</td><td>&quot;Uncertain significance&quot;</td><td>&quot;C4551772:Galloway-Mowat syndro…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>5</td><td>&quot;Pathogenic&quot;</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>6</td><td>&quot;Pathogenic&quot;</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3900693</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900694</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900695</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900696</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900697</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (36_148, 5)\n",
       "┌──────────────┬────────────────────────┬───────────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance   ┆ ReportedPhenotypeInfo     ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                    ┆ ---                       ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                    ┆ str                       ┆ str              ┆ str       │\n",
       "╞══════════════╪════════════════════════╪═══════════════════════════╪══════════════════╪═══════════╡\n",
       "│ 2            ┆ Pathogenic             ┆ C3150901:Hereditary       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ spastic pa…               ┆                  ┆           │\n",
       "│ 3            ┆ Pathogenic             ┆ C3150901:Hereditary       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ spastic pa…               ┆                  ┆           │\n",
       "│ 4            ┆ Uncertain significance ┆ C4551772:Galloway-Mowat   ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ syndro…                   ┆                  ┆           │\n",
       "│ 5            ┆ Pathogenic             ┆ C4748791:Mitochondrial    ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ complex…                  ┆                  ┆           │\n",
       "│ 6            ┆ Pathogenic             ┆ C4748791:Mitochondrial    ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ complex…                  ┆                  ┆           │\n",
       "│ …            ┆ …                      ┆ …                         ┆ …                ┆ …         │\n",
       "│ 3900693      ┆ Pathogenic             ┆ na:GUILLOUET-GORDON       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ SYNDROME                  ┆                  ┆           │\n",
       "│ 3900694      ┆ Pathogenic             ┆ na:GUILLOUET-GORDON       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ SYNDROME                  ┆                  ┆           │\n",
       "│ 3900695      ┆ Pathogenic             ┆ na:GUILLOUET-GORDON       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ SYNDROME                  ┆                  ┆           │\n",
       "│ 3900696      ┆ Pathogenic             ┆ na:GUILLOUET-GORDON       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ SYNDROME                  ┆                  ┆           │\n",
       "│ 3900697      ┆ Pathogenic             ┆ na:GUILLOUET-GORDON       ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                        ┆ SYNDROME                  ┆                  ┆           │\n",
       "└──────────────┴────────────────────────┴───────────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unambiguous_ids = df.group_by(\"#VariationID\").n_unique().filter(ClinicalSignificance=1)[\"#VariationID\"]\n",
    "df = df.filter(pl.col(\"#VariationID\").is_in(unambiguous_ids))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "99a8b8b0-49a3-4671-af89-0772d3a2ea73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (32_910, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;Pathogenic&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3</td><td>&quot;Pathogenic&quot;</td><td>&quot;C3150901:Hereditary spastic pa…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>5</td><td>&quot;Pathogenic&quot;</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>6</td><td>&quot;Pathogenic&quot;</td><td>&quot;C4748791:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>7</td><td>&quot;Pathogenic&quot;</td><td>&quot;C4748792:Mitochondrial complex…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3900693</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900694</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900695</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900696</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3900697</td><td>&quot;Pathogenic&quot;</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (32_910, 5)\n",
       "┌──────────────┬──────────────────────┬─────────────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance ┆ ReportedPhenotypeInfo       ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                  ┆ ---                         ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                  ┆ str                         ┆ str              ┆ str       │\n",
       "╞══════════════╪══════════════════════╪═════════════════════════════╪══════════════════╪═══════════╡\n",
       "│ 2            ┆ Pathogenic           ┆ C3150901:Hereditary spastic ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ pa…                         ┆                  ┆           │\n",
       "│ 3            ┆ Pathogenic           ┆ C3150901:Hereditary spastic ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ pa…                         ┆                  ┆           │\n",
       "│ 5            ┆ Pathogenic           ┆ C4748791:Mitochondrial      ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ complex…                    ┆                  ┆           │\n",
       "│ 6            ┆ Pathogenic           ┆ C4748791:Mitochondrial      ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ complex…                    ┆                  ┆           │\n",
       "│ 7            ┆ Pathogenic           ┆ C4748792:Mitochondrial      ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ complex…                    ┆                  ┆           │\n",
       "│ …            ┆ …                    ┆ …                           ┆ …                ┆ …         │\n",
       "│ 3900693      ┆ Pathogenic           ┆ na:GUILLOUET-GORDON         ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ SYNDROME                    ┆                  ┆           │\n",
       "│ 3900694      ┆ Pathogenic           ┆ na:GUILLOUET-GORDON         ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ SYNDROME                    ┆                  ┆           │\n",
       "│ 3900695      ┆ Pathogenic           ┆ na:GUILLOUET-GORDON         ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ SYNDROME                    ┆                  ┆           │\n",
       "│ 3900696      ┆ Pathogenic           ┆ na:GUILLOUET-GORDON         ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ SYNDROME                    ┆                  ┆           │\n",
       "│ 3900697      ┆ Pathogenic           ┆ na:GUILLOUET-GORDON         ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ SYNDROME                    ┆                  ┆           │\n",
       "└──────────────┴──────────────────────┴─────────────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.filter(pl.col(\"ClinicalSignificance\") == \"Pathogenic\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "93140384",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (0, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (0, 5)\n",
       "┌──────────────┬──────────────────────┬───────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance ┆ ReportedPhenotypeInfo ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                  ┆ ---                   ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                  ┆ str                   ┆ str              ┆ str       │\n",
       "╞══════════════╪══════════════════════╪═══════════════════════╪══════════════════╪═══════════╡\n",
       "└──────────────┴──────────────────────┴───────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"#VariationID\") == 15587)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "19926eb4-384c-446f-a815-650d07f1bd84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: should check that the ClinicalSignificance is the one from OMIM, not the aggregate one from ClinVar\n",
    "# yes, I think it's working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33209854-7e33-432c-8f8a-4e0bd9f1ac36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>4907</td><td>&quot;Pathogenic&quot;</td><td>&quot;C1868114:Polydactyly of a trip…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 5)\n",
       "┌──────────────┬──────────────────────┬─────────────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance ┆ ReportedPhenotypeInfo       ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                  ┆ ---                         ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                  ┆ str                         ┆ str              ┆ str       │\n",
       "╞══════════════╪══════════════════════╪═════════════════════════════╪══════════════════╪═══════════╡\n",
       "│ 4907         ┆ Pathogenic           ┆ C1868114:Polydactyly of a   ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ trip…                       ┆                  ┆           │\n",
       "└──────────────┴──────────────────────┴─────────────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"#VariationID\") == 4907)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8107d1d9-e5fe-4e94-a23e-1aece8ce0b42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>3370406</td><td>&quot;Pathogenic&quot;</td><td>&quot;C5975503:Spermatogenic failure…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr><tr><td>3370406</td><td>&quot;Pathogenic&quot;</td><td>&quot;C5975510:Premature ovarian fai…</td><td>&quot;literature only&quot;</td><td>&quot;OMIM&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2, 5)\n",
       "┌──────────────┬──────────────────────┬─────────────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance ┆ ReportedPhenotypeInfo       ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                  ┆ ---                         ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                  ┆ str                         ┆ str              ┆ str       │\n",
       "╞══════════════╪══════════════════════╪═════════════════════════════╪══════════════════╪═══════════╡\n",
       "│ 3370406      ┆ Pathogenic           ┆ C5975503:Spermatogenic      ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ failure…                    ┆                  ┆           │\n",
       "│ 3370406      ┆ Pathogenic           ┆ C5975510:Premature ovarian  ┆ literature only  ┆ OMIM      │\n",
       "│              ┆                      ┆ fai…                        ┆                  ┆           │\n",
       "└──────────────┴──────────────────────┴─────────────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"#VariationID\") == 3370406)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6bdd3610-068d-4745-9017-895d20195111",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>CollectionMethod</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;literature only&quot;</td><td>32910</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 2)\n",
       "┌──────────────────┬───────┐\n",
       "│ CollectionMethod ┆ count │\n",
       "│ ---              ┆ ---   │\n",
       "│ str              ┆ u32   │\n",
       "╞══════════════════╪═══════╡\n",
       "│ literature only  ┆ 32910 │\n",
       "└──────────────────┴───────┘"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"CollectionMethod\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "783c35a2-4da6-4fd1-8ea0-b7ae7ff902cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# comma is not good, need an escape character, maybe \"|\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0daea2ce-2fab-4b94-bf9c-c458e06d42e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (0, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ClinicalSignificance</th><th>ReportedPhenotypeInfo</th><th>CollectionMethod</th><th>Submitter</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (0, 5)\n",
       "┌──────────────┬──────────────────────┬───────────────────────┬──────────────────┬───────────┐\n",
       "│ #VariationID ┆ ClinicalSignificance ┆ ReportedPhenotypeInfo ┆ CollectionMethod ┆ Submitter │\n",
       "│ ---          ┆ ---                  ┆ ---                   ┆ ---              ┆ ---       │\n",
       "│ i64          ┆ str                  ┆ str                   ┆ str              ┆ str       │\n",
       "╞══════════════╪══════════════════════╪═══════════════════════╪══════════════════╪═══════════╡\n",
       "└──────────────┴──────────────────────┴───────────────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"ReportedPhenotypeInfo\").str.contains(\"|\", literal=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2a29c66b-d0d2-481e-a47a-cbe8053e40f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (31_677, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ReportedPhenotypeInfo</th></tr><tr><td>i64</td><td>str</td></tr></thead><tbody><tr><td>8212</td><td>&quot;CN376810:Basal cell nevus synd…</td></tr><tr><td>17650</td><td>&quot;C2608084:Epidermolysis bullosa…</td></tr><tr><td>126532</td><td>&quot;C3810313:Dowling-Degos disease…</td></tr><tr><td>30705</td><td>&quot;C3280598:Asphyxiating thoracic…</td></tr><tr><td>6610</td><td>&quot;C1832425:Autosomal dominant no…</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>1809806</td><td>&quot;C5975436:Microphthalmia/colobo…</td></tr><tr><td>6374</td><td>&quot;C1842362:Hermansky-Pudlak synd…</td></tr><tr><td>599033</td><td>&quot;C1854469:Noonan syndrome 2&quot;</td></tr><tr><td>976791</td><td>&quot;C4225252:Immunodeficiency 45&quot;</td></tr><tr><td>9226</td><td>&quot;C4551768:Cerebral arteriopathy…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (31_677, 2)\n",
       "┌──────────────┬─────────────────────────────────┐\n",
       "│ #VariationID ┆ ReportedPhenotypeInfo           │\n",
       "│ ---          ┆ ---                             │\n",
       "│ i64          ┆ str                             │\n",
       "╞══════════════╪═════════════════════════════════╡\n",
       "│ 8212         ┆ CN376810:Basal cell nevus synd… │\n",
       "│ 17650        ┆ C2608084:Epidermolysis bullosa… │\n",
       "│ 126532       ┆ C3810313:Dowling-Degos disease… │\n",
       "│ 30705        ┆ C3280598:Asphyxiating thoracic… │\n",
       "│ 6610         ┆ C1832425:Autosomal dominant no… │\n",
       "│ …            ┆ …                               │\n",
       "│ 1809806      ┆ C5975436:Microphthalmia/colobo… │\n",
       "│ 6374         ┆ C1842362:Hermansky-Pudlak synd… │\n",
       "│ 599033       ┆ C1854469:Noonan syndrome 2      │\n",
       "│ 976791       ┆ C4225252:Immunodeficiency 45    │\n",
       "│ 9226         ┆ C4551768:Cerebral arteriopathy… │\n",
       "└──────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.group_by(\"#VariationID\").agg(pl.col(\"ReportedPhenotypeInfo\").unique()).with_columns(pl.col(\"ReportedPhenotypeInfo\").list.join(\"|\"))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7fbdfb3-3b21-4f49-9b4b-812035719a59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1_062, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>#VariationID</th><th>ReportedPhenotypeInfo</th></tr><tr><td>i64</td><td>str</td></tr></thead><tbody><tr><td>14488</td><td>&quot;C2750285:Hutchinson-Gilford pr…</td></tr><tr><td>16331</td><td>&quot;C1300257:Thanatophoric dysplas…</td></tr><tr><td>7479</td><td>&quot;C1847823:Charcot-Marie-Tooth d…</td></tr><tr><td>11609</td><td>&quot;C4551968:Lissencephaly type 1 …</td></tr><tr><td>31146</td><td>&quot;C4706676:Leukoencephalopathy, …</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>18331</td><td>&quot;C3150682:Left ventricular nonc…</td></tr><tr><td>80</td><td>&quot;C0008384:Cholesteryl ester sto…</td></tr><tr><td>285045</td><td>&quot;C1835931:Combined immunodefici…</td></tr><tr><td>15937</td><td>&quot;C2675383:Polyostotic fibrous d…</td></tr><tr><td>5740</td><td>&quot;C1838647:Retinitis pigmentosa …</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1_062, 2)\n",
       "┌──────────────┬─────────────────────────────────┐\n",
       "│ #VariationID ┆ ReportedPhenotypeInfo           │\n",
       "│ ---          ┆ ---                             │\n",
       "│ i64          ┆ str                             │\n",
       "╞══════════════╪═════════════════════════════════╡\n",
       "│ 14488        ┆ C2750285:Hutchinson-Gilford pr… │\n",
       "│ 16331        ┆ C1300257:Thanatophoric dysplas… │\n",
       "│ 7479         ┆ C1847823:Charcot-Marie-Tooth d… │\n",
       "│ 11609        ┆ C4551968:Lissencephaly type 1 … │\n",
       "│ 31146        ┆ C4706676:Leukoencephalopathy, … │\n",
       "│ …            ┆ …                               │\n",
       "│ 18331        ┆ C3150682:Left ventricular nonc… │\n",
       "│ 80           ┆ C0008384:Cholesteryl ester sto… │\n",
       "│ 285045       ┆ C1835931:Combined immunodefici… │\n",
       "│ 15937        ┆ C2675383:Polyostotic fibrous d… │\n",
       "│ 5740         ┆ C1838647:Retinitis pigmentosa … │\n",
       "└──────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.filter(pl.col(\"ReportedPhenotypeInfo\").str.contains(\"|\", literal=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dee49db2-352d-4b54-9ee6-ba02a22866d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (31_677, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>clinvar_id</th><th>ReportedPhenotypeInfo</th></tr><tr><td>i64</td><td>str</td></tr></thead><tbody><tr><td>8212</td><td>&quot;CN376810:Basal cell nevus synd…</td></tr><tr><td>17650</td><td>&quot;C2608084:Epidermolysis bullosa…</td></tr><tr><td>126532</td><td>&quot;C3810313:Dowling-Degos disease…</td></tr><tr><td>30705</td><td>&quot;C3280598:Asphyxiating thoracic…</td></tr><tr><td>6610</td><td>&quot;C1832425:Autosomal dominant no…</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>1809806</td><td>&quot;C5975436:Microphthalmia/colobo…</td></tr><tr><td>6374</td><td>&quot;C1842362:Hermansky-Pudlak synd…</td></tr><tr><td>599033</td><td>&quot;C1854469:Noonan syndrome 2&quot;</td></tr><tr><td>976791</td><td>&quot;C4225252:Immunodeficiency 45&quot;</td></tr><tr><td>9226</td><td>&quot;C4551768:Cerebral arteriopathy…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (31_677, 2)\n",
       "┌────────────┬─────────────────────────────────┐\n",
       "│ clinvar_id ┆ ReportedPhenotypeInfo           │\n",
       "│ ---        ┆ ---                             │\n",
       "│ i64        ┆ str                             │\n",
       "╞════════════╪═════════════════════════════════╡\n",
       "│ 8212       ┆ CN376810:Basal cell nevus synd… │\n",
       "│ 17650      ┆ C2608084:Epidermolysis bullosa… │\n",
       "│ 126532     ┆ C3810313:Dowling-Degos disease… │\n",
       "│ 30705      ┆ C3280598:Asphyxiating thoracic… │\n",
       "│ 6610       ┆ C1832425:Autosomal dominant no… │\n",
       "│ …          ┆ …                               │\n",
       "│ 1809806    ┆ C5975436:Microphthalmia/colobo… │\n",
       "│ 6374       ┆ C1842362:Hermansky-Pudlak synd… │\n",
       "│ 599033     ┆ C1854469:Noonan syndrome 2      │\n",
       "│ 976791     ┆ C4225252:Immunodeficiency 45    │\n",
       "│ 9226       ┆ C4551768:Cerebral arteriopathy… │\n",
       "└────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.rename({\"#VariationID\": \"clinvar_id\"})\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3e4b9bd6-33a4-4471-a482-1d61015c1cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!wget https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/weekly/clinvar_20250601.vcf.gz -O clinvar.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3caf2ece",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.1\n",
      "##fileDate=2025-06-01\n",
      "##source=ClinVar\n",
      "##reference=GRCh38\n",
      "##ID=<Description=\"ClinVar Variation ID\">\n",
      "##INFO=<ID=AF_ESP,Number=1,Type=Float,Description=\"allele frequencies from GO-ESP\">\n",
      "##INFO=<ID=AF_EXAC,Number=1,Type=Float,Description=\"allele frequencies from ExAC\">\n",
      "##INFO=<ID=AF_TGP,Number=1,Type=Float,Description=\"allele frequencies from TGP\">\n",
      "##INFO=<ID=ALLELEID,Number=1,Type=Integer,Description=\"the ClinVar Allele ID\">\n",
      "##INFO=<ID=CLNDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
      "##INFO=<ID=CLNDNINCL,Number=.,Type=String,Description=\"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
      "##INFO=<ID=CLNDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for germline classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=CLNDISDBINCL,Number=.,Type=String,Description=\"For included Variant: Tag-value pairs of disease database name and identifier for germline classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=CLNHGVS,Number=.,Type=String,Description=\"Top-level (primary assembly, alt, or patch) HGVS expression.\">\n",
      "##INFO=<ID=CLNREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of germline classification for the Variation ID\">\n",
      "##INFO=<ID=CLNSIG,Number=.,Type=String,Description=\"Aggregate germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNSIGCONF,Number=.,Type=String,Description=\"Conflicting germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNSIGINCL,Number=.,Type=String,Description=\"Germline classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNSIGSCV,Number=.,Type=String,Description=\"SCV accession numbers for the submissions that contribute to the aggregate germline classification in ClinVar; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=CLNVC,Number=1,Type=String,Description=\"Variant type\">\n",
      "##INFO=<ID=CLNVCSO,Number=1,Type=String,Description=\"Sequence Ontology id for variant type\">\n",
      "##INFO=<ID=CLNVI,Number=.,Type=String,Description=\"the variant's clinical sources reported as tag-value pairs of database and variant identifier\">\n",
      "##INFO=<ID=DBVARID,Number=.,Type=String,Description=\"nsv accessions from dbVar for the variant\">\n",
      "##INFO=<ID=GENEINFO,Number=1,Type=String,Description=\"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\">\n",
      "##INFO=<ID=MC,Number=.,Type=String,Description=\"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\">\n",
      "##INFO=<ID=ONCDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDB\">\n",
      "##INFO=<ID=ONCDNINCL,Number=.,Type=String,Description=\"For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDBINCL\">\n",
      "##INFO=<ID=ONCDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for oncogenicity classifications, e.g. MedGen:NNNNNN\">\n",
      "##INFO=<ID=ONCDISDBINCL,Number=.,Type=String,Description=\"For included variant: Tag-value pairs of disease database name and identifier for oncogenicity classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=ONC,Number=.,Type=String,Description=\"Aggregate oncogenicity classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=ONCINCL,Number=.,Type=String,Description=\"Oncogenicity classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=ONCREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of oncogenicity classification for the Variation ID\">\n",
      "##INFO=<ID=ONCSCV,Number=.,Type=String,Description=\"SCV accession numbers for the submissions that contribute to the aggregate oncogenicity classification in ClinVar; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=ONCCONF,Number=.,Type=String,Description=\"Conflicting oncogenicity classification for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=ORIGIN,Number=.,Type=String,Description=\"Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other\">\n",
      "##INFO=<ID=RS,Number=.,Type=String,Description=\"dbSNP ID (i.e. rs number)\">\n",
      "##INFO=<ID=SCIDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDB\">\n",
      "##INFO=<ID=SCIDNINCL,Number=.,Type=String,Description=\"For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDBINCL\">\n",
      "##INFO=<ID=SCIDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for somatic clinial impact classifications, e.g. MedGen:NNNNNN\">\n",
      "##INFO=<ID=SCIDISDBINCL,Number=.,Type=String,Description=\"For included variant: Tag-value pairs of disease database name and identifier for somatic clinical impact classifications, e.g. OMIM:NNNNNN\">\n",
      "##INFO=<ID=SCIREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of somatic clinical impact for the Variation ID\">\n",
      "##INFO=<ID=SCI,Number=.,Type=String,Description=\"Aggregate somatic clinical impact for this single variant; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=SCIINCL,Number=.,Type=String,Description=\"Somatic clinical impact classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
      "##INFO=<ID=SCISCV,Number=.,Type=String,Description=\"SCV accession numbers for the submissions that contribute to the aggregate somatic clinical impact in ClinVar; multiple values are separated by a vertical bar\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
      "1\t66926\t3385321\tAG\tA\t.\t.\tALLELEID=3544463;CLNDISDB=Human_Phenotype_Ontology:HP:0000547,MONDO:MONDO:0019200,MeSH:D012174,MedGen:C0035334,OMIM:268000,OMIM:PS268000,Orphanet:791;CLNDN=Retinitis_pigmentosa;CLNHGVS=NC_000001.11:g.66927del;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNSIGSCV=SCV005419006;CLNVC=Deletion;CLNVCSO=SO:0000159;GENEINFO=OR4F5:79501;MC=SO:0001627|intron_variant;ORIGIN=0\n",
      "1\t69134\t2205837\tA\tG\t.\t.\tALLELEID=2193183;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69134A>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNSIGSCV=SCV003526545;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=ClinGen:CA502008;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1;RS=781394307\n",
      "1\t69314\t3205580\tT\tG\t.\t.\tALLELEID=3374047;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69314T>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNSIGSCV=SCV004995495;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=ClinGen:CA338197388;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1;RS=2521653848\n",
      "1\t69423\t3205581\tG\tA\t.\t.\tALLELEID=3374048;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69423G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNSIGSCV=SCV004995496;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=ClinGen:CA338197763;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1;RS=2521654033\n",
      "1\t69581\t2252161\tC\tG\t.\t.\tALLELEID=2238986;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69581C>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNSIGSCV=SCV003584698;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=ClinGen:CA338198277;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1570409925\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat clinvar.vcf.gz | head -n 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "23cee897-c883-4acd-84a5-2b67c81b9e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "NUCLEOTIDES = list(\"ACGT\")\n",
    "CHROMS = [str(i) for i in range(1, 23)] + ['X', 'Y']\n",
    "COORDINATES = [\"chrom\", \"pos\", \"ref\", \"alt\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f73d18a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'Y' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'MT' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113889.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187633.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187661.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187693.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_009646201.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_235_717, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>69134</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>2205837</td></tr><tr><td>&quot;1&quot;</td><td>69314</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>3205580</td></tr><tr><td>&quot;1&quot;</td><td>69423</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>3205581</td></tr><tr><td>&quot;1&quot;</td><td>69581</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>2252161</td></tr><tr><td>&quot;1&quot;</td><td>69682</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>2396347</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;NT_187693.1&quot;</td><td>273806</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>2219599</td></tr><tr><td>&quot;NT_187693.1&quot;</td><td>273866</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>2237818</td></tr><tr><td>&quot;NT_187693.1&quot;</td><td>274185</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>3778023</td></tr><tr><td>&quot;NT_187693.1&quot;</td><td>274366</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>2206666</td></tr><tr><td>&quot;NT_187693.1&quot;</td><td>275068</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>2241971</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_235_717, 5)\n",
       "┌─────────────┬────────┬─────┬─────┬────────────┐\n",
       "│ chrom       ┆ pos    ┆ ref ┆ alt ┆ clinvar_id │\n",
       "│ ---         ┆ ---    ┆ --- ┆ --- ┆ ---        │\n",
       "│ str         ┆ i64    ┆ str ┆ str ┆ i64        │\n",
       "╞═════════════╪════════╪═════╪═════╪════════════╡\n",
       "│ 1           ┆ 69134  ┆ A   ┆ G   ┆ 2205837    │\n",
       "│ 1           ┆ 69314  ┆ T   ┆ G   ┆ 3205580    │\n",
       "│ 1           ┆ 69423  ┆ G   ┆ A   ┆ 3205581    │\n",
       "│ 1           ┆ 69581  ┆ C   ┆ G   ┆ 2252161    │\n",
       "│ 1           ┆ 69682  ┆ G   ┆ A   ┆ 2396347    │\n",
       "│ …           ┆ …      ┆ …   ┆ …   ┆ …          │\n",
       "│ NT_187693.1 ┆ 273806 ┆ G   ┆ A   ┆ 2219599    │\n",
       "│ NT_187693.1 ┆ 273866 ┆ A   ┆ C   ┆ 2237818    │\n",
       "│ NT_187693.1 ┆ 274185 ┆ C   ┆ T   ┆ 3778023    │\n",
       "│ NT_187693.1 ┆ 274366 ┆ G   ┆ C   ┆ 2206666    │\n",
       "│ NT_187693.1 ┆ 275068 ┆ T   ┆ C   ┆ 2241971    │\n",
       "└─────────────┴────────┴─────┴─────┴────────────┘"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cyvcf2 import VCF\n",
    "\n",
    "rows = []\n",
    "for variant in VCF(\"clinvar.vcf.gz\"):\n",
    "    if variant.INFO.get(\"CLNVC\") != \"single_nucleotide_variant\": continue\n",
    "    if len(variant.ALT) != 1:\n",
    "        continue\n",
    "    rows.append([variant.CHROM, variant.POS, variant.REF, variant.ALT[0], int(variant.ID)])\n",
    "\n",
    "V = pl.DataFrame(rows, schema=[\"chrom\", \"pos\", \"ref\", \"alt\", \"clinvar_id\"], orient=\"row\")\n",
    "V"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "53e68802-f21f-49b5-a438-781d7b502767",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_232_804, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>69134</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>2205837</td></tr><tr><td>&quot;1&quot;</td><td>69314</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>3205580</td></tr><tr><td>&quot;1&quot;</td><td>69423</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>3205581</td></tr><tr><td>&quot;1&quot;</td><td>69581</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>2252161</td></tr><tr><td>&quot;1&quot;</td><td>69682</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>2396347</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Y&quot;</td><td>14830121</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>391879</td></tr><tr><td>&quot;Y&quot;</td><td>14840423</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>2689591</td></tr><tr><td>&quot;Y&quot;</td><td>14840785</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>770316</td></tr><tr><td>&quot;Y&quot;</td><td>14840887</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>2661892</td></tr><tr><td>&quot;Y&quot;</td><td>21548157</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>2689863</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_232_804, 5)\n",
       "┌───────┬──────────┬─────┬─────┬────────────┐\n",
       "│ chrom ┆ pos      ┆ ref ┆ alt ┆ clinvar_id │\n",
       "│ ---   ┆ ---      ┆ --- ┆ --- ┆ ---        │\n",
       "│ str   ┆ i64      ┆ str ┆ str ┆ i64        │\n",
       "╞═══════╪══════════╪═════╪═════╪════════════╡\n",
       "│ 1     ┆ 69134    ┆ A   ┆ G   ┆ 2205837    │\n",
       "│ 1     ┆ 69314    ┆ T   ┆ G   ┆ 3205580    │\n",
       "│ 1     ┆ 69423    ┆ G   ┆ A   ┆ 3205581    │\n",
       "│ 1     ┆ 69581    ┆ C   ┆ G   ┆ 2252161    │\n",
       "│ 1     ┆ 69682    ┆ G   ┆ A   ┆ 2396347    │\n",
       "│ …     ┆ …        ┆ …   ┆ …   ┆ …          │\n",
       "│ Y     ┆ 14830121 ┆ C   ┆ A   ┆ 391879     │\n",
       "│ Y     ┆ 14840423 ┆ C   ┆ T   ┆ 2689591    │\n",
       "│ Y     ┆ 14840785 ┆ C   ┆ T   ┆ 770316     │\n",
       "│ Y     ┆ 14840887 ┆ C   ┆ T   ┆ 2661892    │\n",
       "│ Y     ┆ 21548157 ┆ G   ┆ T   ┆ 2689863    │\n",
       "└───────┴──────────┴─────┴─────┴────────────┘"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V = V.filter(pl.col(\"chrom\").is_in(CHROMS), pl.col(\"ref\").is_in(NUCLEOTIDES), pl.col(\"alt\").is_in(NUCLEOTIDES))\n",
    "V"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd3f42f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8_276, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>clinvar_id</th><th>ReportedPhenotypeInfo</th></tr><tr><td>i64</td><td>str</td></tr></thead><tbody><tr><td>2</td><td>&quot;C3150901:Hereditary spastic pa…</td></tr><tr><td>3</td><td>&quot;C3150901:Hereditary spastic pa…</td></tr><tr><td>7</td><td>&quot;C4748792:Mitochondrial complex…</td></tr><tr><td>22</td><td>&quot;C3150874:Cranioectodermal dysp…</td></tr><tr><td>24</td><td>&quot;C2675204:PHARC syndrome&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3900693</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td></tr><tr><td>3900694</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td></tr><tr><td>3900695</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td></tr><tr><td>3900696</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td></tr><tr><td>3900697</td><td>&quot;na:GUILLOUET-GORDON SYNDROME&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8_276, 2)\n",
       "┌────────────┬─────────────────────────────────┐\n",
       "│ clinvar_id ┆ ReportedPhenotypeInfo           │\n",
       "│ ---        ┆ ---                             │\n",
       "│ i64        ┆ str                             │\n",
       "╞════════════╪═════════════════════════════════╡\n",
       "│ 2          ┆ C3150901:Hereditary spastic pa… │\n",
       "│ 3          ┆ C3150901:Hereditary spastic pa… │\n",
       "│ 7          ┆ C4748792:Mitochondrial complex… │\n",
       "│ 22         ┆ C3150874:Cranioectodermal dysp… │\n",
       "│ 24         ┆ C2675204:PHARC syndrome         │\n",
       "│ …          ┆ …                               │\n",
       "│ 3900693    ┆ na:GUILLOUET-GORDON SYNDROME    │\n",
       "│ 3900694    ┆ na:GUILLOUET-GORDON SYNDROME    │\n",
       "│ 3900695    ┆ na:GUILLOUET-GORDON SYNDROME    │\n",
       "│ 3900696    ┆ na:GUILLOUET-GORDON SYNDROME    │\n",
       "│ 3900697    ┆ na:GUILLOUET-GORDON SYNDROME    │\n",
       "└────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maybe they are not single nucleotide variants\n",
    "# 2: indel\n",
    "# 3: indel\n",
    "# 7: seems to be a haplotype (2 linked variants)\n",
    "# 22: indel\n",
    "# 24: indel\n",
    "# 3900693: another chromosome (NM_005481.3)\n",
    "df.join(V, on=\"clinvar_id\", how=\"anti\").sort(\"clinvar_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d49adc18-5fdc-4d91-a9a6-a478db201cba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (23_401, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th><th>ReportedPhenotypeInfo</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>1014143</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>183381</td><td>&quot;C4015293:Mendelian susceptibil…</td></tr><tr><td>&quot;1&quot;</td><td>1014359</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>161454</td><td>&quot;C4015293:Mendelian susceptibil…</td></tr><tr><td>&quot;1&quot;</td><td>1041582</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>126556</td><td>&quot;C3808739:Congenital myasthenic…</td></tr><tr><td>&quot;1&quot;</td><td>1050575</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>18241</td><td>&quot;C3808739:Congenital myasthenic…</td></tr><tr><td>&quot;1&quot;</td><td>1050763</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>126555</td><td>&quot;C3808739:Congenital myasthenic…</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Y&quot;</td><td>2787426</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>9739</td><td>&quot;C2748896:46,XY sex reversal 1&quot;</td></tr><tr><td>&quot;Y&quot;</td><td>2787551</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>9754</td><td>&quot;C2748896:46,XY sex reversal 1&quot;</td></tr><tr><td>&quot;Y&quot;</td><td>2787592</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>9751</td><td>&quot;C2748896:46,XY sex reversal 1&quot;</td></tr><tr><td>&quot;Y&quot;</td><td>2787600</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>9753</td><td>&quot;C2748896:46,XY sex reversal 1&quot;</td></tr><tr><td>&quot;Y&quot;</td><td>7063898</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>625467</td><td>&quot;C5193013:Deafness, Y-linked 2&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (23_401, 6)\n",
       "┌───────┬─────────┬─────┬─────┬────────────┬─────────────────────────────────┐\n",
       "│ chrom ┆ pos     ┆ ref ┆ alt ┆ clinvar_id ┆ ReportedPhenotypeInfo           │\n",
       "│ ---   ┆ ---     ┆ --- ┆ --- ┆ ---        ┆ ---                             │\n",
       "│ str   ┆ i64     ┆ str ┆ str ┆ i64        ┆ str                             │\n",
       "╞═══════╪═════════╪═════╪═════╪════════════╪═════════════════════════════════╡\n",
       "│ 1     ┆ 1014143 ┆ C   ┆ T   ┆ 183381     ┆ C4015293:Mendelian susceptibil… │\n",
       "│ 1     ┆ 1014359 ┆ G   ┆ T   ┆ 161454     ┆ C4015293:Mendelian susceptibil… │\n",
       "│ 1     ┆ 1041582 ┆ C   ┆ T   ┆ 126556     ┆ C3808739:Congenital myasthenic… │\n",
       "│ 1     ┆ 1050575 ┆ G   ┆ C   ┆ 18241      ┆ C3808739:Congenital myasthenic… │\n",
       "│ 1     ┆ 1050763 ┆ G   ┆ T   ┆ 126555     ┆ C3808739:Congenital myasthenic… │\n",
       "│ …     ┆ …       ┆ …   ┆ …   ┆ …          ┆ …                               │\n",
       "│ Y     ┆ 2787426 ┆ C   ┆ G   ┆ 9739       ┆ C2748896:46,XY sex reversal 1   │\n",
       "│ Y     ┆ 2787551 ┆ C   ┆ T   ┆ 9754       ┆ C2748896:46,XY sex reversal 1   │\n",
       "│ Y     ┆ 2787592 ┆ A   ┆ T   ┆ 9751       ┆ C2748896:46,XY sex reversal 1   │\n",
       "│ Y     ┆ 2787600 ┆ G   ┆ A   ┆ 9753       ┆ C2748896:46,XY sex reversal 1   │\n",
       "│ Y     ┆ 7063898 ┆ A   ┆ T   ┆ 625467     ┆ C5193013:Deafness, Y-linked 2   │\n",
       "└───────┴─────────┴─────┴─────┴────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V = V.join(df, how=\"inner\", on=\"clinvar_id\")\n",
    "V"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "51cf8194",
   "metadata": {},
   "outputs": [],
   "source": [
    "#V.write_parquet(\"clinvar_omim_variants_v1.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e0a057d4-7a3f-4690-acfa-e18171112912",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23401, 31677)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(V), len(df)  # TODO: we are dropping too many, need to understand why\n",
    "# start by analyzing some specific examples, are there in clinvar database?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4468cb54-dd87-4043-b697-d2a3f0fc19b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th><th>ReportedPhenotypeInfo</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;7&quot;</td><td>156791480</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>4907</td><td>&quot;C1868114:Polydactyly of a trip…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 6)\n",
       "┌───────┬───────────┬─────┬─────┬────────────┬─────────────────────────────────┐\n",
       "│ chrom ┆ pos       ┆ ref ┆ alt ┆ clinvar_id ┆ ReportedPhenotypeInfo           │\n",
       "│ ---   ┆ ---       ┆ --- ┆ --- ┆ ---        ┆ ---                             │\n",
       "│ str   ┆ i64       ┆ str ┆ str ┆ i64        ┆ str                             │\n",
       "╞═══════╪═══════════╪═════╪═════╪════════════╪═════════════════════════════════╡\n",
       "│ 7     ┆ 156791480 ┆ G   ┆ A   ┆ 4907       ┆ C1868114:Polydactyly of a trip… │\n",
       "└───────┴───────────┴─────┴─────┴────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V.filter(clinvar_id=4907)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "061e6dfc-42ba-4323-99e5-86716331ad04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (338, 9)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>OMIM</th><th>consequence</th><th>label</th><th>tss_dist</th><th>match_group</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>7961859</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>&quot;MIM 606324&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>34</td><td>&quot;PLS_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>9943502</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 608553&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>26</td><td>&quot;5_prime_UTR_variant_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>9943503</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 608553&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>27</td><td>&quot;5_prime_UTR_variant_1&quot;</td></tr><tr><td>&quot;1&quot;</td><td>11023351</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;MIM 612069&quot;</td><td>&quot;3_prime_UTR_variant&quot;</td><td>true</td><td>1206</td><td>&quot;3_prime_UTR_variant_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>21509427</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 241500&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>0</td><td>&quot;5_prime_UTR_variant_2&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;X&quot;</td><td>155022770</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>46</td><td>&quot;PLS_57&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022771</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>47</td><td>&quot;PLS_62&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022773</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>49</td><td>&quot;PLS_58&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022807</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>83</td><td>&quot;PLS_59&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022809</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>85</td><td>&quot;PLS_60&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (338, 9)\n",
       "┌───────┬───────────┬─────┬─────┬───┬─────────────────────┬───────┬──────────┬─────────────────────┐\n",
       "│ chrom ┆ pos       ┆ ref ┆ alt ┆ … ┆ consequence         ┆ label ┆ tss_dist ┆ match_group         │\n",
       "│ ---   ┆ ---       ┆ --- ┆ --- ┆   ┆ ---                 ┆ ---   ┆ ---      ┆ ---                 │\n",
       "│ str   ┆ i64       ┆ str ┆ str ┆   ┆ str                 ┆ bool  ┆ i64      ┆ str                 │\n",
       "╞═══════╪═══════════╪═════╪═════╪═══╪═════════════════════╪═══════╪══════════╪═════════════════════╡\n",
       "│ 1     ┆ 7961859   ┆ C   ┆ G   ┆ … ┆ PLS                 ┆ true  ┆ 34       ┆ PLS_0               │\n",
       "│ 1     ┆ 9943502   ┆ A   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 26       ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _0                  │\n",
       "│ 1     ┆ 9943503   ┆ C   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 27       ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _1                  │\n",
       "│ 1     ┆ 11023351  ┆ G   ┆ A   ┆ … ┆ 3_prime_UTR_variant ┆ true  ┆ 1206     ┆ 3_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _0                  │\n",
       "│ 1     ┆ 21509427  ┆ C   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 0        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _2                  │\n",
       "│ …     ┆ …         ┆ …   ┆ …   ┆ … ┆ …                   ┆ …     ┆ …        ┆ …                   │\n",
       "│ X     ┆ 155022770 ┆ A   ┆ G   ┆ … ┆ PLS                 ┆ true  ┆ 46       ┆ PLS_57              │\n",
       "│ X     ┆ 155022771 ┆ G   ┆ A   ┆ … ┆ PLS                 ┆ true  ┆ 47       ┆ PLS_62              │\n",
       "│ X     ┆ 155022773 ┆ A   ┆ T   ┆ … ┆ PLS                 ┆ true  ┆ 49       ┆ PLS_58              │\n",
       "│ X     ┆ 155022807 ┆ T   ┆ C   ┆ … ┆ PLS                 ┆ true  ┆ 83       ┆ PLS_59              │\n",
       "│ X     ┆ 155022809 ┆ A   ┆ C   ┆ … ┆ PLS                 ┆ true  ┆ 85       ┆ PLS_60              │\n",
       "└───────┴───────────┴─────┴─────┴───┴─────────────────────┴───────┴──────────┴─────────────────────┘"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traitgym = pl.read_parquet('hf://datasets/songlab/TraitGym/mendelian_traits_matched_9/test.parquet').filter(pl.col(\"label\"))\n",
    "traitgym"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "291f7e20-8922-450e-ad23-7894629b786c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (120, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th><th>ReportedPhenotypeInfo</th><th>OMIM</th><th>consequence</th><th>label</th><th>tss_dist</th><th>match_group</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>11023351</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>5239</td><td>&quot;C3148872:FRONTOTEMPORAL DEMENT…</td><td>&quot;MIM 612069&quot;</td><td>&quot;3_prime_UTR_variant&quot;</td><td>true</td><td>1206</td><td>&quot;3_prime_UTR_variant_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>90916206</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>31109</td><td>&quot;C3279997:Myopia 21, autosomal …</td><td>&quot;MIM 614167&quot;</td><td>&quot;3_prime_UTR_variant&quot;</td><td>true</td><td>105250</td><td>&quot;3_prime_UTR_variant_3&quot;</td></tr><tr><td>&quot;1&quot;</td><td>112956192</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>8916</td><td>&quot;C1864902:Exercise-induced hype…</td><td>&quot;MIM 610021&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>3</td><td>&quot;5_prime_UTR_variant_5&quot;</td></tr><tr><td>&quot;1&quot;</td><td>155301478</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>1515</td><td>&quot;C0340968:Pyruvate kinase defic…</td><td>&quot;MIM 266200&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>39</td><td>&quot;PLS_2&quot;</td></tr><tr><td>&quot;1&quot;</td><td>160032009</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>1288</td><td>&quot;C5201145:Hypercoagulability sy…</td><td>&quot;MIM 610293&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>18</td><td>&quot;PLS_3&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;X&quot;</td><td>139530730</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>641767</td><td>&quot;C5848256:Hemophilia B leyden&quot;</td><td>&quot;MIM 306900&quot;</td><td>&quot;upstream_gene_variant&quot;</td><td>true</td><td>8</td><td>&quot;upstream_gene_variant_17&quot;</td></tr><tr><td>&quot;X&quot;</td><td>139530731</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>10645</td><td>&quot;C5848256:Hemophilia B leyden&quot;</td><td>&quot;MIM 306900&quot;</td><td>&quot;upstream_gene_variant&quot;</td><td>true</td><td>7</td><td>&quot;upstream_gene_variant_19&quot;</td></tr><tr><td>&quot;X&quot;</td><td>139530743</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>10644</td><td>&quot;C5848256:Hemophilia B leyden&quot;</td><td>&quot;MIM 306900&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>3</td><td>&quot;5_prime_UTR_variant_107&quot;</td></tr><tr><td>&quot;X&quot;</td><td>139530748</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>10646</td><td>&quot;C5848256:Hemophilia B leyden&quot;</td><td>&quot;MIM 306900&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>8</td><td>&quot;5_prime_UTR_variant_110&quot;</td></tr><tr><td>&quot;X&quot;</td><td>153971810</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>39694</td><td>&quot;C0796208:Methylmalonic acidemi…</td><td>&quot;MIM 309541&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>7</td><td>&quot;5_prime_UTR_variant_100&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (120, 11)\n",
       "┌───────┬───────────┬─────┬─────┬───┬─────────────────────┬───────┬──────────┬─────────────────────┐\n",
       "│ chrom ┆ pos       ┆ ref ┆ alt ┆ … ┆ consequence         ┆ label ┆ tss_dist ┆ match_group         │\n",
       "│ ---   ┆ ---       ┆ --- ┆ --- ┆   ┆ ---                 ┆ ---   ┆ ---      ┆ ---                 │\n",
       "│ str   ┆ i64       ┆ str ┆ str ┆   ┆ str                 ┆ bool  ┆ i64      ┆ str                 │\n",
       "╞═══════╪═══════════╪═════╪═════╪═══╪═════════════════════╪═══════╪══════════╪═════════════════════╡\n",
       "│ 1     ┆ 11023351  ┆ G   ┆ A   ┆ … ┆ 3_prime_UTR_variant ┆ true  ┆ 1206     ┆ 3_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _0                  │\n",
       "│ 1     ┆ 90916206  ┆ C   ┆ T   ┆ … ┆ 3_prime_UTR_variant ┆ true  ┆ 105250   ┆ 3_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _3                  │\n",
       "│ 1     ┆ 112956192 ┆ C   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 3        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _5                  │\n",
       "│ 1     ┆ 155301478 ┆ C   ┆ G   ┆ … ┆ PLS                 ┆ true  ┆ 39       ┆ PLS_2               │\n",
       "│ 1     ┆ 160032009 ┆ G   ┆ C   ┆ … ┆ PLS                 ┆ true  ┆ 18       ┆ PLS_3               │\n",
       "│ …     ┆ …         ┆ …   ┆ …   ┆ … ┆ …                   ┆ …     ┆ …        ┆ …                   │\n",
       "│ X     ┆ 139530730 ┆ G   ┆ A   ┆ … ┆ upstream_gene_varia ┆ true  ┆ 8        ┆ upstream_gene_varia │\n",
       "│       ┆           ┆     ┆     ┆   ┆ nt                  ┆       ┆          ┆ nt_17               │\n",
       "│ X     ┆ 139530731 ┆ A   ┆ T   ┆ … ┆ upstream_gene_varia ┆ true  ┆ 7        ┆ upstream_gene_varia │\n",
       "│       ┆           ┆     ┆     ┆   ┆ nt                  ┆       ┆          ┆ nt_19               │\n",
       "│ X     ┆ 139530743 ┆ T   ┆ C   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 3        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _107                │\n",
       "│ X     ┆ 139530748 ┆ A   ┆ G   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 8        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _110                │\n",
       "│ X     ┆ 153971810 ┆ A   ┆ G   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 7        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _100                │\n",
       "└───────┴───────────┴─────┴─────┴───┴─────────────────────┴───────┴──────────┴─────────────────────┘"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = V.join(traitgym, how=\"inner\", on=COORDINATES)  # dropping too many, seems random though, no obvious pattern\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "58f1e266-e80f-4cd6-8348-cd56ea2972b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (9, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>consequence</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;pELS_flank&quot;</td><td>3</td></tr><tr><td>&quot;5_prime_UTR_variant&quot;</td><td>41</td></tr><tr><td>&quot;dELS&quot;</td><td>9</td></tr><tr><td>&quot;dELS_flank&quot;</td><td>3</td></tr><tr><td>&quot;3_prime_UTR_variant&quot;</td><td>11</td></tr><tr><td>&quot;PLS&quot;</td><td>16</td></tr><tr><td>&quot;upstream_gene_variant&quot;</td><td>5</td></tr><tr><td>&quot;intron_variant&quot;</td><td>1</td></tr><tr><td>&quot;non_coding_transcript_exon_var…</td><td>31</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (9, 2)\n",
       "┌─────────────────────────────────┬───────┐\n",
       "│ consequence                     ┆ count │\n",
       "│ ---                             ┆ ---   │\n",
       "│ str                             ┆ u32   │\n",
       "╞═════════════════════════════════╪═══════╡\n",
       "│ pELS_flank                      ┆ 3     │\n",
       "│ 5_prime_UTR_variant             ┆ 41    │\n",
       "│ dELS                            ┆ 9     │\n",
       "│ dELS_flank                      ┆ 3     │\n",
       "│ 3_prime_UTR_variant             ┆ 11    │\n",
       "│ PLS                             ┆ 16    │\n",
       "│ upstream_gene_variant           ┆ 5     │\n",
       "│ intron_variant                  ┆ 1     │\n",
       "│ non_coding_transcript_exon_var… ┆ 31    │\n",
       "└─────────────────────────────────┴───────┘"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x[\"consequence\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4997533e-a627-4123-830b-41b701b6464b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (218, 9)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>OMIM</th><th>consequence</th><th>label</th><th>tss_dist</th><th>match_group</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>7961859</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>&quot;MIM 606324&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>34</td><td>&quot;PLS_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>9943502</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 608553&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>26</td><td>&quot;5_prime_UTR_variant_0&quot;</td></tr><tr><td>&quot;1&quot;</td><td>9943503</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 608553&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>27</td><td>&quot;5_prime_UTR_variant_1&quot;</td></tr><tr><td>&quot;1&quot;</td><td>21509427</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 241500&quot;</td><td>&quot;5_prime_UTR_variant&quot;</td><td>true</td><td>0</td><td>&quot;5_prime_UTR_variant_2&quot;</td></tr><tr><td>&quot;1&quot;</td><td>25816825</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>&quot;MIM 602771&quot;</td><td>&quot;3_prime_UTR_variant&quot;</td><td>true</td><td>1814</td><td>&quot;3_prime_UTR_variant_1&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;X&quot;</td><td>155022770</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>46</td><td>&quot;PLS_57&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022771</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>47</td><td>&quot;PLS_62&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022773</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>49</td><td>&quot;PLS_58&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022807</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>83</td><td>&quot;PLS_59&quot;</td></tr><tr><td>&quot;X&quot;</td><td>155022809</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>&quot;MIM 306700&quot;</td><td>&quot;PLS&quot;</td><td>true</td><td>85</td><td>&quot;PLS_60&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (218, 9)\n",
       "┌───────┬───────────┬─────┬─────┬───┬─────────────────────┬───────┬──────────┬─────────────────────┐\n",
       "│ chrom ┆ pos       ┆ ref ┆ alt ┆ … ┆ consequence         ┆ label ┆ tss_dist ┆ match_group         │\n",
       "│ ---   ┆ ---       ┆ --- ┆ --- ┆   ┆ ---                 ┆ ---   ┆ ---      ┆ ---                 │\n",
       "│ str   ┆ i64       ┆ str ┆ str ┆   ┆ str                 ┆ bool  ┆ i64      ┆ str                 │\n",
       "╞═══════╪═══════════╪═════╪═════╪═══╪═════════════════════╪═══════╪══════════╪═════════════════════╡\n",
       "│ 1     ┆ 7961859   ┆ C   ┆ G   ┆ … ┆ PLS                 ┆ true  ┆ 34       ┆ PLS_0               │\n",
       "│ 1     ┆ 9943502   ┆ A   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 26       ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _0                  │\n",
       "│ 1     ┆ 9943503   ┆ C   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 27       ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _1                  │\n",
       "│ 1     ┆ 21509427  ┆ C   ┆ T   ┆ … ┆ 5_prime_UTR_variant ┆ true  ┆ 0        ┆ 5_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _2                  │\n",
       "│ 1     ┆ 25816825  ┆ T   ┆ C   ┆ … ┆ 3_prime_UTR_variant ┆ true  ┆ 1814     ┆ 3_prime_UTR_variant │\n",
       "│       ┆           ┆     ┆     ┆   ┆                     ┆       ┆          ┆ _1                  │\n",
       "│ …     ┆ …         ┆ …   ┆ …   ┆ … ┆ …                   ┆ …     ┆ …        ┆ …                   │\n",
       "│ X     ┆ 155022770 ┆ A   ┆ G   ┆ … ┆ PLS                 ┆ true  ┆ 46       ┆ PLS_57              │\n",
       "│ X     ┆ 155022771 ┆ G   ┆ A   ┆ … ┆ PLS                 ┆ true  ┆ 47       ┆ PLS_62              │\n",
       "│ X     ┆ 155022773 ┆ A   ┆ T   ┆ … ┆ PLS                 ┆ true  ┆ 49       ┆ PLS_58              │\n",
       "│ X     ┆ 155022807 ┆ T   ┆ C   ┆ … ┆ PLS                 ┆ true  ┆ 83       ┆ PLS_59              │\n",
       "│ X     ┆ 155022809 ┆ A   ┆ C   ┆ … ┆ PLS                 ┆ true  ┆ 85       ┆ PLS_60              │\n",
       "└───────┴───────────┴─────┴─────┴───┴─────────────────────┴───────┴──────────┴─────────────────────┘"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maybe some are no longer pathogenic according to OMIM?\n",
    "# or maybe, Smedley et al. did their own literature review rather than take them from OMIM\n",
    "# TODO: check if these missing variants are perhaps not in UCSC genome browser OMIM\n",
    "traitgym.join(V, how=\"anti\", on=COORDINATES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1d72f0b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_232_804, 5)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>pos</th><th>ref</th><th>alt</th><th>clinvar_id</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td></tr></thead><tbody><tr><td>&quot;1&quot;</td><td>69134</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>2205837</td></tr><tr><td>&quot;1&quot;</td><td>69314</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>3205580</td></tr><tr><td>&quot;1&quot;</td><td>69423</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>3205581</td></tr><tr><td>&quot;1&quot;</td><td>69581</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>2252161</td></tr><tr><td>&quot;1&quot;</td><td>69682</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>2396347</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Y&quot;</td><td>14830121</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>391879</td></tr><tr><td>&quot;Y&quot;</td><td>14840423</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>2689591</td></tr><tr><td>&quot;Y&quot;</td><td>14840785</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>770316</td></tr><tr><td>&quot;Y&quot;</td><td>14840887</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>2661892</td></tr><tr><td>&quot;Y&quot;</td><td>21548157</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>2689863</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_232_804, 5)\n",
       "┌───────┬──────────┬─────┬─────┬────────────┐\n",
       "│ chrom ┆ pos      ┆ ref ┆ alt ┆ clinvar_id │\n",
       "│ ---   ┆ ---      ┆ --- ┆ --- ┆ ---        │\n",
       "│ str   ┆ i64      ┆ str ┆ str ┆ i64        │\n",
       "╞═══════╪══════════╪═════╪═════╪════════════╡\n",
       "│ 1     ┆ 69134    ┆ A   ┆ G   ┆ 2205837    │\n",
       "│ 1     ┆ 69314    ┆ T   ┆ G   ┆ 3205580    │\n",
       "│ 1     ┆ 69423    ┆ G   ┆ A   ┆ 3205581    │\n",
       "│ 1     ┆ 69581    ┆ C   ┆ G   ┆ 2252161    │\n",
       "│ 1     ┆ 69682    ┆ G   ┆ A   ┆ 2396347    │\n",
       "│ …     ┆ …        ┆ …   ┆ …   ┆ …          │\n",
       "│ Y     ┆ 14830121 ┆ C   ┆ A   ┆ 391879     │\n",
       "│ Y     ┆ 14840423 ┆ C   ┆ T   ┆ 2689591    │\n",
       "│ Y     ┆ 14840785 ┆ C   ┆ T   ┆ 770316     │\n",
       "│ Y     ┆ 14840887 ┆ C   ┆ T   ┆ 2661892    │\n",
       "│ Y     ┆ 21548157 ┆ G   ┆ T   ┆ 2689863    │\n",
       "└───────┴──────────┴─────┴─────┴────────────┘"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d126656",
   "metadata": {},
   "source": [
    "# After uploading to Ensembl VEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df4535b5-574a-4d80-8ba5-d9100a8058c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>PUBMED</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>TRANSCRIPTION_FACTORS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1014143_C/T</td>\n",
       "      <td>1:1014143-1014143</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1014359_G/T</td>\n",
       "      <td>1:1014359-1014359</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1041582_C/T</td>\n",
       "      <td>1:1041582-1041582</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1050575_G/C</td>\n",
       "      <td>1:1050575-1050575</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1050763_G/T</td>\n",
       "      <td>1:1050763-1050763</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23396</th>\n",
       "      <td>Y_2787426_C/G</td>\n",
       "      <td>Y:2787426-2787426</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23397</th>\n",
       "      <td>Y_2787551_C/T</td>\n",
       "      <td>Y:2787551-2787551</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23398</th>\n",
       "      <td>Y_2787592_A/T</td>\n",
       "      <td>Y:2787592-2787592</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23399</th>\n",
       "      <td>Y_2787600_G/A</td>\n",
       "      <td>Y:2787600-2787600</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23400</th>\n",
       "      <td>Y_7063898_A/T</td>\n",
       "      <td>Y:7063898-7063898</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23401 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation           Location Allele       Consequence IMPACT  \\\n",
       "0           1_1014143_C/T  1:1014143-1014143      -       stop_gained      -   \n",
       "1           1_1014359_G/T  1:1014359-1014359      -       stop_gained      -   \n",
       "2           1_1041582_C/T  1:1041582-1041582      -       stop_gained      -   \n",
       "3           1_1050575_G/C  1:1050575-1050575      -  missense_variant      -   \n",
       "4           1_1050763_G/T  1:1050763-1050763      -  missense_variant      -   \n",
       "...                   ...                ...    ...               ...    ...   \n",
       "23396       Y_2787426_C/G  Y:2787426-2787426      -  missense_variant      -   \n",
       "23397       Y_2787551_C/T  Y:2787551-2787551      -  missense_variant      -   \n",
       "23398       Y_2787592_A/T  Y:2787592-2787592      -       stop_gained      -   \n",
       "23399       Y_2787600_G/A  Y:2787600-2787600      -       stop_gained      -   \n",
       "23400       Y_7063898_A/T  Y:7063898-7063898      -  missense_variant      -   \n",
       "\n",
       "      SYMBOL Gene Feature_type Feature BIOTYPE  ... AF CLIN_SIG SOMATIC PHENO  \\\n",
       "0          -    -            -       -       -  ...  -        -       -     -   \n",
       "1          -    -            -       -       -  ...  -        -       -     -   \n",
       "2          -    -            -       -       -  ...  -        -       -     -   \n",
       "3          -    -            -       -       -  ...  -        -       -     -   \n",
       "4          -    -            -       -       -  ...  -        -       -     -   \n",
       "...      ...  ...          ...     ...     ...  ... ..      ...     ...   ...   \n",
       "23396      -    -            -       -       -  ...  -        -       -     -   \n",
       "23397      -    -            -       -       -  ...  -        -       -     -   \n",
       "23398      -    -            -       -       -  ...  -        -       -     -   \n",
       "23399      -    -            -       -       -  ...  -        -       -     -   \n",
       "23400      -    -            -       -       -  ...  -        -       -     -   \n",
       "\n",
       "      PUBMED MOTIF_NAME MOTIF_POS HIGH_INF_POS MOTIF_SCORE_CHANGE  \\\n",
       "0          -          -         -            -                  -   \n",
       "1          -          -         -            -                  -   \n",
       "2          -          -         -            -                  -   \n",
       "3          -          -         -            -                  -   \n",
       "4          -          -         -            -                  -   \n",
       "...      ...        ...       ...          ...                ...   \n",
       "23396      -          -         -            -                  -   \n",
       "23397      -          -         -            -                  -   \n",
       "23398      -          -         -            -                  -   \n",
       "23399      -          -         -            -                  -   \n",
       "23400      -          -         -            -                  -   \n",
       "\n",
       "      TRANSCRIPTION_FACTORS  \n",
       "0                         -  \n",
       "1                         -  \n",
       "2                         -  \n",
       "3                         -  \n",
       "4                         -  \n",
       "...                     ...  \n",
       "23396                     -  \n",
       "23397                     -  \n",
       "23398                     -  \n",
       "23399                     -  \n",
       "23400                     -  \n",
       "\n",
       "[23401 rows x 37 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "V2 = pd.read_csv(\"XG4kWDtTujjeGoxJ.txt\", sep=\"\\t\")\n",
    "V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "502ba928",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>PUBMED</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>TRANSCRIPTION_FACTORS</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1014143_C/T</td>\n",
       "      <td>1:1014143-1014143</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1014359_G/T</td>\n",
       "      <td>1:1014359-1014359</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1041582_C/T</td>\n",
       "      <td>1:1041582-1041582</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1050575_G/C</td>\n",
       "      <td>1:1050575-1050575</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1050575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1050763_G/T</td>\n",
       "      <td>1:1050763-1050763</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1050763</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23396</th>\n",
       "      <td>Y_2787426_C/G</td>\n",
       "      <td>Y:2787426-2787426</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>2787426</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23397</th>\n",
       "      <td>Y_2787551_C/T</td>\n",
       "      <td>Y:2787551-2787551</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>2787551</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23398</th>\n",
       "      <td>Y_2787592_A/T</td>\n",
       "      <td>Y:2787592-2787592</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>2787592</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23399</th>\n",
       "      <td>Y_2787600_G/A</td>\n",
       "      <td>Y:2787600-2787600</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>2787600</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23400</th>\n",
       "      <td>Y_7063898_A/T</td>\n",
       "      <td>Y:7063898-7063898</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>7063898</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23401 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation           Location Allele       consequence IMPACT  \\\n",
       "0           1_1014143_C/T  1:1014143-1014143      -       stop_gained      -   \n",
       "1           1_1014359_G/T  1:1014359-1014359      -       stop_gained      -   \n",
       "2           1_1041582_C/T  1:1041582-1041582      -       stop_gained      -   \n",
       "3           1_1050575_G/C  1:1050575-1050575      -  missense_variant      -   \n",
       "4           1_1050763_G/T  1:1050763-1050763      -  missense_variant      -   \n",
       "...                   ...                ...    ...               ...    ...   \n",
       "23396       Y_2787426_C/G  Y:2787426-2787426      -  missense_variant      -   \n",
       "23397       Y_2787551_C/T  Y:2787551-2787551      -  missense_variant      -   \n",
       "23398       Y_2787592_A/T  Y:2787592-2787592      -       stop_gained      -   \n",
       "23399       Y_2787600_G/A  Y:2787600-2787600      -       stop_gained      -   \n",
       "23400       Y_7063898_A/T  Y:7063898-7063898      -  missense_variant      -   \n",
       "\n",
       "      SYMBOL Gene Feature_type Feature BIOTYPE  ... PUBMED MOTIF_NAME  \\\n",
       "0          -    -            -       -       -  ...      -          -   \n",
       "1          -    -            -       -       -  ...      -          -   \n",
       "2          -    -            -       -       -  ...      -          -   \n",
       "3          -    -            -       -       -  ...      -          -   \n",
       "4          -    -            -       -       -  ...      -          -   \n",
       "...      ...  ...          ...     ...     ...  ...    ...        ...   \n",
       "23396      -    -            -       -       -  ...      -          -   \n",
       "23397      -    -            -       -       -  ...      -          -   \n",
       "23398      -    -            -       -       -  ...      -          -   \n",
       "23399      -    -            -       -       -  ...      -          -   \n",
       "23400      -    -            -       -       -  ...      -          -   \n",
       "\n",
       "      MOTIF_POS HIGH_INF_POS MOTIF_SCORE_CHANGE TRANSCRIPTION_FACTORS chrom  \\\n",
       "0             -            -                  -                     -     1   \n",
       "1             -            -                  -                     -     1   \n",
       "2             -            -                  -                     -     1   \n",
       "3             -            -                  -                     -     1   \n",
       "4             -            -                  -                     -     1   \n",
       "...         ...          ...                ...                   ...   ...   \n",
       "23396         -            -                  -                     -     Y   \n",
       "23397         -            -                  -                     -     Y   \n",
       "23398         -            -                  -                     -     Y   \n",
       "23399         -            -                  -                     -     Y   \n",
       "23400         -            -                  -                     -     Y   \n",
       "\n",
       "           pos ref alt  \n",
       "0      1014143   C   T  \n",
       "1      1014359   G   T  \n",
       "2      1041582   C   T  \n",
       "3      1050575   G   C  \n",
       "4      1050763   G   T  \n",
       "...        ...  ..  ..  \n",
       "23396  2787426   C   G  \n",
       "23397  2787551   C   T  \n",
       "23398  2787592   A   T  \n",
       "23399  2787600   G   A  \n",
       "23400  7063898   A   T  \n",
       "\n",
       "[23401 rows x 41 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[\"chrom\"] = V2[\"#Uploaded_variation\"].str.split(\"_\").str[0]\n",
    "V2[\"pos\"] = V2[\"#Uploaded_variation\"].str.split(\"_\").str[1].astype(int)\n",
    "V2[\"ref\"] = V2[\"#Uploaded_variation\"].str.split(\"_\").str[2].str.split(\"/\").str[0]\n",
    "V2[\"alt\"] = V2[\"#Uploaded_variation\"].str.split(\"_\").str[2].str.split(\"/\").str[1]\n",
    "V2.rename(columns={\"Consequence\": \"consequence\"}, inplace=True)\n",
    "V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6493f4c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "V2[\"start\"] = V2.pos - 1\n",
    "V2[\"end\"] = V2.pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "04659b52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12697</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674100</th>\n",
       "      <td>Y</td>\n",
       "      <td>26628270</td>\n",
       "      <td>26628437</td>\n",
       "      <td>ENSG00000237917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674101</th>\n",
       "      <td>Y</td>\n",
       "      <td>26630646</td>\n",
       "      <td>26630749</td>\n",
       "      <td>ENSG00000237917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674102</th>\n",
       "      <td>Y</td>\n",
       "      <td>26633344</td>\n",
       "      <td>26633431</td>\n",
       "      <td>ENSG00000237917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674103</th>\n",
       "      <td>Y</td>\n",
       "      <td>26634522</td>\n",
       "      <td>26634652</td>\n",
       "      <td>ENSG00000237917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674104</th>\n",
       "      <td>Y</td>\n",
       "      <td>56855243</td>\n",
       "      <td>56855488</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>674105 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom     start       end          gene_id\n",
       "0          1     11868     12227  ENSG00000223972\n",
       "1          1     12009     12057  ENSG00000223972\n",
       "2          1     12178     12227  ENSG00000223972\n",
       "3          1     12612     12697  ENSG00000223972\n",
       "4          1     12612     12721  ENSG00000223972\n",
       "...      ...       ...       ...              ...\n",
       "674100     Y  26628270  26628437  ENSG00000237917\n",
       "674101     Y  26630646  26630749  ENSG00000237917\n",
       "674102     Y  26633344  26633431  ENSG00000237917\n",
       "674103     Y  26634522  26634652  ENSG00000237917\n",
       "674104     Y  56855243  56855488  ENSG00000235857\n",
       "\n",
       "[674105 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon = pd.read_parquet(\"../../results/exon.parquet\")\n",
    "exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f9af70b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/users/gbenegas/software/mambaforge/envs/gpn/lib/python3.11/site-packages/bioframe/core/specs.py:150: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  pd.api.types.is_categorical_dtype(chrom_dtype),\n",
      "/scratch/users/gbenegas/software/mambaforge/envs/gpn/lib/python3.11/site-packages/bioframe/core/specs.py:150: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  pd.api.types.is_categorical_dtype(chrom_dtype),\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>gene_id_</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1014143_C/T</td>\n",
       "      <td>1:1014143-1014143</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1014143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014435</td>\n",
       "      <td>ENSG00000187608</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1014359_G/T</td>\n",
       "      <td>1:1014359-1014359</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>1</td>\n",
       "      <td>1013983</td>\n",
       "      <td>1014435</td>\n",
       "      <td>ENSG00000187608</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1041582_C/T</td>\n",
       "      <td>1:1041582-1041582</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1041582</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>1</td>\n",
       "      <td>1041477</td>\n",
       "      <td>1041702</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1050575_G/C</td>\n",
       "      <td>1:1050575-1050575</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1050575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1050574</td>\n",
       "      <td>1050575</td>\n",
       "      <td>1</td>\n",
       "      <td>1050426</td>\n",
       "      <td>1050591</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1050763_G/T</td>\n",
       "      <td>1:1050763-1050763</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1050763</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1050762</td>\n",
       "      <td>1050763</td>\n",
       "      <td>1</td>\n",
       "      <td>1050725</td>\n",
       "      <td>1050837</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23396</th>\n",
       "      <td>Y_2787426_C/G</td>\n",
       "      <td>Y:2787426-2787426</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2787426</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2787425</td>\n",
       "      <td>2787426</td>\n",
       "      <td>Y</td>\n",
       "      <td>2786854</td>\n",
       "      <td>2787682</td>\n",
       "      <td>ENSG00000184895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23397</th>\n",
       "      <td>Y_2787551_C/T</td>\n",
       "      <td>Y:2787551-2787551</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2787551</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2787550</td>\n",
       "      <td>2787551</td>\n",
       "      <td>Y</td>\n",
       "      <td>2786854</td>\n",
       "      <td>2787682</td>\n",
       "      <td>ENSG00000184895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23398</th>\n",
       "      <td>Y_2787592_A/T</td>\n",
       "      <td>Y:2787592-2787592</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2787592</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2787591</td>\n",
       "      <td>2787592</td>\n",
       "      <td>Y</td>\n",
       "      <td>2786854</td>\n",
       "      <td>2787682</td>\n",
       "      <td>ENSG00000184895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23399</th>\n",
       "      <td>Y_2787600_G/A</td>\n",
       "      <td>Y:2787600-2787600</td>\n",
       "      <td>-</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2787600</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2787599</td>\n",
       "      <td>2787600</td>\n",
       "      <td>Y</td>\n",
       "      <td>2786854</td>\n",
       "      <td>2787682</td>\n",
       "      <td>ENSG00000184895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23400</th>\n",
       "      <td>Y_7063898_A/T</td>\n",
       "      <td>Y:7063898-7063898</td>\n",
       "      <td>-</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>7063898</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>7063897</td>\n",
       "      <td>7063898</td>\n",
       "      <td>Y</td>\n",
       "      <td>7063896</td>\n",
       "      <td>7064149</td>\n",
       "      <td>ENSG00000092377</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23401 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation           Location Allele       consequence IMPACT  \\\n",
       "0           1_1014143_C/T  1:1014143-1014143      -       stop_gained      -   \n",
       "1           1_1014359_G/T  1:1014359-1014359      -       stop_gained      -   \n",
       "2           1_1041582_C/T  1:1041582-1041582      -       stop_gained      -   \n",
       "3           1_1050575_G/C  1:1050575-1050575      -  missense_variant      -   \n",
       "4           1_1050763_G/T  1:1050763-1050763      -  missense_variant      -   \n",
       "...                   ...                ...    ...               ...    ...   \n",
       "23396       Y_2787426_C/G  Y:2787426-2787426      -  missense_variant      -   \n",
       "23397       Y_2787551_C/T  Y:2787551-2787551      -  missense_variant      -   \n",
       "23398       Y_2787592_A/T  Y:2787592-2787592      -       stop_gained      -   \n",
       "23399       Y_2787600_G/A  Y:2787600-2787600      -       stop_gained      -   \n",
       "23400       Y_7063898_A/T  Y:7063898-7063898      -  missense_variant      -   \n",
       "\n",
       "      SYMBOL Gene Feature_type Feature BIOTYPE  ...      pos ref alt    start  \\\n",
       "0          -    -            -       -       -  ...  1014143   C   T  1014142   \n",
       "1          -    -            -       -       -  ...  1014359   G   T  1014358   \n",
       "2          -    -            -       -       -  ...  1041582   C   T  1041581   \n",
       "3          -    -            -       -       -  ...  1050575   G   C  1050574   \n",
       "4          -    -            -       -       -  ...  1050763   G   T  1050762   \n",
       "...      ...  ...          ...     ...     ...  ...      ...  ..  ..      ...   \n",
       "23396      -    -            -       -       -  ...  2787426   C   G  2787425   \n",
       "23397      -    -            -       -       -  ...  2787551   C   T  2787550   \n",
       "23398      -    -            -       -       -  ...  2787592   A   T  2787591   \n",
       "23399      -    -            -       -       -  ...  2787600   G   A  2787599   \n",
       "23400      -    -            -       -       -  ...  7063898   A   T  7063897   \n",
       "\n",
       "           end chrom_   start_     end_         gene_id_ distance  \n",
       "0      1014143      1  1013983  1014435  ENSG00000187608        0  \n",
       "1      1014359      1  1013983  1014435  ENSG00000187608        0  \n",
       "2      1041582      1  1041477  1041702  ENSG00000188157        0  \n",
       "3      1050575      1  1050426  1050591  ENSG00000188157        0  \n",
       "4      1050763      1  1050725  1050837  ENSG00000188157        0  \n",
       "...        ...    ...      ...      ...              ...      ...  \n",
       "23396  2787426      Y  2786854  2787682  ENSG00000184895        0  \n",
       "23397  2787551      Y  2786854  2787682  ENSG00000184895        0  \n",
       "23398  2787592      Y  2786854  2787682  ENSG00000184895        0  \n",
       "23399  2787600      Y  2786854  2787682  ENSG00000184895        0  \n",
       "23400  7063898      Y  7063896  7064149  ENSG00000092377        0  \n",
       "\n",
       "[23401 rows x 48 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import bioframe as bf\n",
    "\n",
    "V2 = bf.closest(V2, exon)\n",
    "V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "586b75a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count          149.0\n",
       "mean     1083.228188\n",
       "std      2873.273686\n",
       "min              0.0\n",
       "25%             25.0\n",
       "50%             87.0\n",
       "75%            705.0\n",
       "max          22818.0\n",
       "Name: distance, dtype: Float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[V2.consequence == \"intron_variant\"].distance.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "832f13c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>gene_id_</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19868</th>\n",
       "      <td>7_156791472_C/G</td>\n",
       "      <td>7:156791472-156791472</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791472</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>156791471</td>\n",
       "      <td>156791472</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19869</th>\n",
       "      <td>7_156791472_C/T</td>\n",
       "      <td>7:156791472-156791472</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791472</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>156791471</td>\n",
       "      <td>156791472</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation               Location Allele     consequence  \\\n",
       "19868     7_156791472_C/G  7:156791472-156791472      -  intron_variant   \n",
       "19869     7_156791472_C/T  7:156791472-156791472      -  intron_variant   \n",
       "\n",
       "      IMPACT SYMBOL Gene Feature_type Feature BIOTYPE  ...        pos ref alt  \\\n",
       "19868      -      -    -            -       -       -  ...  156791472   C   G   \n",
       "19869      -      -    -            -       -       -  ...  156791472   C   T   \n",
       "\n",
       "           start        end chrom_     start_       end_         gene_id_  \\\n",
       "19868  156791471  156791472      7  156796388  156796493  ENSG00000105983   \n",
       "19869  156791471  156791472      7  156796388  156796493  ENSG00000105983   \n",
       "\n",
       "      distance  \n",
       "19868     4916  \n",
       "19869     4916  \n",
       "\n",
       "[2 rows x 48 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[V2.pos == 156791472]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a5001dfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(V2[V2.consequence == \"intron_variant\"].distance > 1000).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7167b130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>gene_id_</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1880</th>\n",
       "      <td>1_215891198_T/C</td>\n",
       "      <td>1:215891198-215891198</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>215891198</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>215891197</td>\n",
       "      <td>215891198</td>\n",
       "      <td>1</td>\n",
       "      <td>215888425</td>\n",
       "      <td>215889054</td>\n",
       "      <td>ENSG00000042781</td>\n",
       "      <td>2143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5470</th>\n",
       "      <td>12_88101183_T/C</td>\n",
       "      <td>12:88101183-88101183</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>88101183</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>88101182</td>\n",
       "      <td>88101183</td>\n",
       "      <td>12</td>\n",
       "      <td>88102837</td>\n",
       "      <td>88103516</td>\n",
       "      <td>ENSG00000198707</td>\n",
       "      <td>1654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6039</th>\n",
       "      <td>13_48471962_A/G</td>\n",
       "      <td>13:48471962-48471962</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>48471962</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>48471961</td>\n",
       "      <td>48471962</td>\n",
       "      <td>13</td>\n",
       "      <td>48473359</td>\n",
       "      <td>48473390</td>\n",
       "      <td>ENSG00000139687</td>\n",
       "      <td>1397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7958</th>\n",
       "      <td>16_8832245_C/T</td>\n",
       "      <td>16:8832245-8832245</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>8832245</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>8832244</td>\n",
       "      <td>8832245</td>\n",
       "      <td>16</td>\n",
       "      <td>8829192</td>\n",
       "      <td>8829258</td>\n",
       "      <td>ENSG00000140650</td>\n",
       "      <td>2986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11180</th>\n",
       "      <td>19_38583494_A/G</td>\n",
       "      <td>19:38583494-38583494</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>38583494</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>38583493</td>\n",
       "      <td>38583494</td>\n",
       "      <td>19</td>\n",
       "      <td>38584903</td>\n",
       "      <td>38585099</td>\n",
       "      <td>ENSG00000196218</td>\n",
       "      <td>1409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12043</th>\n",
       "      <td>2_62904596_G/A</td>\n",
       "      <td>2:62904596-62904596</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>62904596</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>62904595</td>\n",
       "      <td>62904596</td>\n",
       "      <td>2</td>\n",
       "      <td>62881691</td>\n",
       "      <td>62881777</td>\n",
       "      <td>ENSG00000115504</td>\n",
       "      <td>22818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13010</th>\n",
       "      <td>2_219128549_T/C</td>\n",
       "      <td>2:219128549-219128549</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>219128549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>219128548</td>\n",
       "      <td>219128549</td>\n",
       "      <td>2</td>\n",
       "      <td>219125470</td>\n",
       "      <td>219125660</td>\n",
       "      <td>ENSG00000187736</td>\n",
       "      <td>2888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14894</th>\n",
       "      <td>3_33119323_C/G</td>\n",
       "      <td>3:33119323-33119323</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>33119323</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>33119322</td>\n",
       "      <td>33119323</td>\n",
       "      <td>3</td>\n",
       "      <td>33120343</td>\n",
       "      <td>33120493</td>\n",
       "      <td>ENSG00000170275</td>\n",
       "      <td>1020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15503</th>\n",
       "      <td>3_119515865_A/G</td>\n",
       "      <td>3:119515865-119515865</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>119515865</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>119515864</td>\n",
       "      <td>119515865</td>\n",
       "      <td>3</td>\n",
       "      <td>119517204</td>\n",
       "      <td>119517315</td>\n",
       "      <td>ENSG00000113845</td>\n",
       "      <td>1339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15587</th>\n",
       "      <td>3_124055231_T/G</td>\n",
       "      <td>3:124055231-124055231</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>124055231</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>124055230</td>\n",
       "      <td>124055231</td>\n",
       "      <td>3</td>\n",
       "      <td>124033584</td>\n",
       "      <td>124033813</td>\n",
       "      <td>ENSG00000160145</td>\n",
       "      <td>21417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18378</th>\n",
       "      <td>6_75884246_T/G</td>\n",
       "      <td>6:75884246-75884246</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>75884246</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>75884245</td>\n",
       "      <td>75884246</td>\n",
       "      <td>6</td>\n",
       "      <td>75886003</td>\n",
       "      <td>75886094</td>\n",
       "      <td>ENSG00000196586</td>\n",
       "      <td>1757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19580</th>\n",
       "      <td>7_117639961_C/T</td>\n",
       "      <td>7:117639961-117639961</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>117639961</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>117639960</td>\n",
       "      <td>117639961</td>\n",
       "      <td>7</td>\n",
       "      <td>117642437</td>\n",
       "      <td>117642593</td>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>2476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19865</th>\n",
       "      <td>7_156791137_T/C</td>\n",
       "      <td>7:156791137-156791137</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791137</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>156791136</td>\n",
       "      <td>156791137</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>5251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19866</th>\n",
       "      <td>7_156791255_G/C</td>\n",
       "      <td>7:156791255-156791255</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791255</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>156791254</td>\n",
       "      <td>156791255</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>5133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19867</th>\n",
       "      <td>7_156791470_T/C</td>\n",
       "      <td>7:156791470-156791470</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791470</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>156791469</td>\n",
       "      <td>156791470</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19868</th>\n",
       "      <td>7_156791472_C/G</td>\n",
       "      <td>7:156791472-156791472</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791472</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>156791471</td>\n",
       "      <td>156791472</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19869</th>\n",
       "      <td>7_156791472_C/T</td>\n",
       "      <td>7:156791472-156791472</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791472</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>156791471</td>\n",
       "      <td>156791472</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19870</th>\n",
       "      <td>7_156791474_G/A</td>\n",
       "      <td>7:156791474-156791474</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791474</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>156791473</td>\n",
       "      <td>156791474</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19871</th>\n",
       "      <td>7_156791480_G/A</td>\n",
       "      <td>7:156791480-156791480</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791480</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>156791479</td>\n",
       "      <td>156791480</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19872</th>\n",
       "      <td>7_156791542_A/C</td>\n",
       "      <td>7:156791542-156791542</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791542</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>156791541</td>\n",
       "      <td>156791542</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19873</th>\n",
       "      <td>7_156791547_A/G</td>\n",
       "      <td>7:156791547-156791547</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791547</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>156791546</td>\n",
       "      <td>156791547</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19874</th>\n",
       "      <td>7_156791548_G/C</td>\n",
       "      <td>7:156791548-156791548</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791548</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>156791547</td>\n",
       "      <td>156791548</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19875</th>\n",
       "      <td>7_156791571_T/A</td>\n",
       "      <td>7:156791571-156791571</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791571</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>156791570</td>\n",
       "      <td>156791571</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19876</th>\n",
       "      <td>7_156791579_C/T</td>\n",
       "      <td>7:156791579-156791579</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791579</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>156791578</td>\n",
       "      <td>156791579</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19877</th>\n",
       "      <td>7_156791581_A/G</td>\n",
       "      <td>7:156791581-156791581</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>156791580</td>\n",
       "      <td>156791581</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19878</th>\n",
       "      <td>7_156791771_G/C</td>\n",
       "      <td>7:156791771-156791771</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>156791771</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>156791770</td>\n",
       "      <td>156791771</td>\n",
       "      <td>7</td>\n",
       "      <td>156796388</td>\n",
       "      <td>156796493</td>\n",
       "      <td>ENSG00000105983</td>\n",
       "      <td>4617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21693</th>\n",
       "      <td>X_22095367_G/T</td>\n",
       "      <td>X:22095367-22095367</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>22095367</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>22095366</td>\n",
       "      <td>22095367</td>\n",
       "      <td>X</td>\n",
       "      <td>22093982</td>\n",
       "      <td>22094099</td>\n",
       "      <td>ENSG00000102174</td>\n",
       "      <td>1267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21776</th>\n",
       "      <td>X_32461403_T/C</td>\n",
       "      <td>X:32461403-32461403</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>32461403</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>32461402</td>\n",
       "      <td>32461403</td>\n",
       "      <td>X</td>\n",
       "      <td>32463438</td>\n",
       "      <td>32463594</td>\n",
       "      <td>ENSG00000198947</td>\n",
       "      <td>2035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21800</th>\n",
       "      <td>X_33014549_A/T</td>\n",
       "      <td>X:33014549-33014549</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>33014549</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>33014548</td>\n",
       "      <td>33014549</td>\n",
       "      <td>X</td>\n",
       "      <td>33019306</td>\n",
       "      <td>33020200</td>\n",
       "      <td>ENSG00000198947</td>\n",
       "      <td>4757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23354</th>\n",
       "      <td>X_154991304_C/T</td>\n",
       "      <td>X:154991304-154991304</td>\n",
       "      <td>-</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>154991304</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>154991303</td>\n",
       "      <td>154991304</td>\n",
       "      <td>X</td>\n",
       "      <td>154992935</td>\n",
       "      <td>154993148</td>\n",
       "      <td>ENSG00000185010</td>\n",
       "      <td>1631</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation               Location Allele     consequence  \\\n",
       "1880      1_215891198_T/C  1:215891198-215891198      -  intron_variant   \n",
       "5470      12_88101183_T/C   12:88101183-88101183      -  intron_variant   \n",
       "6039      13_48471962_A/G   13:48471962-48471962      -  intron_variant   \n",
       "7958       16_8832245_C/T     16:8832245-8832245      -  intron_variant   \n",
       "11180     19_38583494_A/G   19:38583494-38583494      -  intron_variant   \n",
       "12043      2_62904596_G/A    2:62904596-62904596      -  intron_variant   \n",
       "13010     2_219128549_T/C  2:219128549-219128549      -  intron_variant   \n",
       "14894      3_33119323_C/G    3:33119323-33119323      -  intron_variant   \n",
       "15503     3_119515865_A/G  3:119515865-119515865      -  intron_variant   \n",
       "15587     3_124055231_T/G  3:124055231-124055231      -  intron_variant   \n",
       "18378      6_75884246_T/G    6:75884246-75884246      -  intron_variant   \n",
       "19580     7_117639961_C/T  7:117639961-117639961      -  intron_variant   \n",
       "19865     7_156791137_T/C  7:156791137-156791137      -  intron_variant   \n",
       "19866     7_156791255_G/C  7:156791255-156791255      -  intron_variant   \n",
       "19867     7_156791470_T/C  7:156791470-156791470      -  intron_variant   \n",
       "19868     7_156791472_C/G  7:156791472-156791472      -  intron_variant   \n",
       "19869     7_156791472_C/T  7:156791472-156791472      -  intron_variant   \n",
       "19870     7_156791474_G/A  7:156791474-156791474      -  intron_variant   \n",
       "19871     7_156791480_G/A  7:156791480-156791480      -  intron_variant   \n",
       "19872     7_156791542_A/C  7:156791542-156791542      -  intron_variant   \n",
       "19873     7_156791547_A/G  7:156791547-156791547      -  intron_variant   \n",
       "19874     7_156791548_G/C  7:156791548-156791548      -  intron_variant   \n",
       "19875     7_156791571_T/A  7:156791571-156791571      -  intron_variant   \n",
       "19876     7_156791579_C/T  7:156791579-156791579      -  intron_variant   \n",
       "19877     7_156791581_A/G  7:156791581-156791581      -  intron_variant   \n",
       "19878     7_156791771_G/C  7:156791771-156791771      -  intron_variant   \n",
       "21693      X_22095367_G/T    X:22095367-22095367      -  intron_variant   \n",
       "21776      X_32461403_T/C    X:32461403-32461403      -  intron_variant   \n",
       "21800      X_33014549_A/T    X:33014549-33014549      -  intron_variant   \n",
       "23354     X_154991304_C/T  X:154991304-154991304      -  intron_variant   \n",
       "\n",
       "      IMPACT SYMBOL Gene Feature_type Feature BIOTYPE  ...        pos ref alt  \\\n",
       "1880       -      -    -            -       -       -  ...  215891198   T   C   \n",
       "5470       -      -    -            -       -       -  ...   88101183   T   C   \n",
       "6039       -      -    -            -       -       -  ...   48471962   A   G   \n",
       "7958       -      -    -            -       -       -  ...    8832245   C   T   \n",
       "11180      -      -    -            -       -       -  ...   38583494   A   G   \n",
       "12043      -      -    -            -       -       -  ...   62904596   G   A   \n",
       "13010      -      -    -            -       -       -  ...  219128549   T   C   \n",
       "14894      -      -    -            -       -       -  ...   33119323   C   G   \n",
       "15503      -      -    -            -       -       -  ...  119515865   A   G   \n",
       "15587      -      -    -            -       -       -  ...  124055231   T   G   \n",
       "18378      -      -    -            -       -       -  ...   75884246   T   G   \n",
       "19580      -      -    -            -       -       -  ...  117639961   C   T   \n",
       "19865      -      -    -            -       -       -  ...  156791137   T   C   \n",
       "19866      -      -    -            -       -       -  ...  156791255   G   C   \n",
       "19867      -      -    -            -       -       -  ...  156791470   T   C   \n",
       "19868      -      -    -            -       -       -  ...  156791472   C   G   \n",
       "19869      -      -    -            -       -       -  ...  156791472   C   T   \n",
       "19870      -      -    -            -       -       -  ...  156791474   G   A   \n",
       "19871      -      -    -            -       -       -  ...  156791480   G   A   \n",
       "19872      -      -    -            -       -       -  ...  156791542   A   C   \n",
       "19873      -      -    -            -       -       -  ...  156791547   A   G   \n",
       "19874      -      -    -            -       -       -  ...  156791548   G   C   \n",
       "19875      -      -    -            -       -       -  ...  156791571   T   A   \n",
       "19876      -      -    -            -       -       -  ...  156791579   C   T   \n",
       "19877      -      -    -            -       -       -  ...  156791581   A   G   \n",
       "19878      -      -    -            -       -       -  ...  156791771   G   C   \n",
       "21693      -      -    -            -       -       -  ...   22095367   G   T   \n",
       "21776      -      -    -            -       -       -  ...   32461403   T   C   \n",
       "21800      -      -    -            -       -       -  ...   33014549   A   T   \n",
       "23354      -      -    -            -       -       -  ...  154991304   C   T   \n",
       "\n",
       "           start        end chrom_     start_       end_         gene_id_  \\\n",
       "1880   215891197  215891198      1  215888425  215889054  ENSG00000042781   \n",
       "5470    88101182   88101183     12   88102837   88103516  ENSG00000198707   \n",
       "6039    48471961   48471962     13   48473359   48473390  ENSG00000139687   \n",
       "7958     8832244    8832245     16    8829192    8829258  ENSG00000140650   \n",
       "11180   38583493   38583494     19   38584903   38585099  ENSG00000196218   \n",
       "12043   62904595   62904596      2   62881691   62881777  ENSG00000115504   \n",
       "13010  219128548  219128549      2  219125470  219125660  ENSG00000187736   \n",
       "14894   33119322   33119323      3   33120343   33120493  ENSG00000170275   \n",
       "15503  119515864  119515865      3  119517204  119517315  ENSG00000113845   \n",
       "15587  124055230  124055231      3  124033584  124033813  ENSG00000160145   \n",
       "18378   75884245   75884246      6   75886003   75886094  ENSG00000196586   \n",
       "19580  117639960  117639961      7  117642437  117642593  ENSG00000001626   \n",
       "19865  156791136  156791137      7  156796388  156796493  ENSG00000105983   \n",
       "19866  156791254  156791255      7  156796388  156796493  ENSG00000105983   \n",
       "19867  156791469  156791470      7  156796388  156796493  ENSG00000105983   \n",
       "19868  156791471  156791472      7  156796388  156796493  ENSG00000105983   \n",
       "19869  156791471  156791472      7  156796388  156796493  ENSG00000105983   \n",
       "19870  156791473  156791474      7  156796388  156796493  ENSG00000105983   \n",
       "19871  156791479  156791480      7  156796388  156796493  ENSG00000105983   \n",
       "19872  156791541  156791542      7  156796388  156796493  ENSG00000105983   \n",
       "19873  156791546  156791547      7  156796388  156796493  ENSG00000105983   \n",
       "19874  156791547  156791548      7  156796388  156796493  ENSG00000105983   \n",
       "19875  156791570  156791571      7  156796388  156796493  ENSG00000105983   \n",
       "19876  156791578  156791579      7  156796388  156796493  ENSG00000105983   \n",
       "19877  156791580  156791581      7  156796388  156796493  ENSG00000105983   \n",
       "19878  156791770  156791771      7  156796388  156796493  ENSG00000105983   \n",
       "21693   22095366   22095367      X   22093982   22094099  ENSG00000102174   \n",
       "21776   32461402   32461403      X   32463438   32463594  ENSG00000198947   \n",
       "21800   33014548   33014549      X   33019306   33020200  ENSG00000198947   \n",
       "23354  154991303  154991304      X  154992935  154993148  ENSG00000185010   \n",
       "\n",
       "      distance  \n",
       "1880      2143  \n",
       "5470      1654  \n",
       "6039      1397  \n",
       "7958      2986  \n",
       "11180     1409  \n",
       "12043    22818  \n",
       "13010     2888  \n",
       "14894     1020  \n",
       "15503     1339  \n",
       "15587    21417  \n",
       "18378     1757  \n",
       "19580     2476  \n",
       "19865     5251  \n",
       "19866     5133  \n",
       "19867     4918  \n",
       "19868     4916  \n",
       "19869     4916  \n",
       "19870     4914  \n",
       "19871     4908  \n",
       "19872     4846  \n",
       "19873     4841  \n",
       "19874     4840  \n",
       "19875     4817  \n",
       "19876     4809  \n",
       "19877     4807  \n",
       "19878     4617  \n",
       "21693     1267  \n",
       "21776     2035  \n",
       "21800     4757  \n",
       "23354     1631  \n",
       "\n",
       "[30 rows x 48 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[(V2.consequence == \"intron_variant\") & (V2.distance > 1000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "265da960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "consequence\n",
       "missense_variant                       15755\n",
       "stop_gained                             4525\n",
       "splice_donor_variant                    1041\n",
       "splice_acceptor_variant                  749\n",
       "start_lost                               208\n",
       "splice_donor_5th_base_variant            189\n",
       "splice_region_variant                    178\n",
       "intron_variant                           149\n",
       "splice_donor_region_variant              147\n",
       "splice_polypyrimidine_tract_variant       84\n",
       "5_prime_UTR_variant                       80\n",
       "non_coding_transcript_exon_variant        79\n",
       "stop_lost                                 70\n",
       "synonymous_variant                        61\n",
       "3_prime_UTR_variant                       39\n",
       "upstream_gene_variant                     33\n",
       "mature_miRNA_variant                       5\n",
       "regulatory_region_variant                  4\n",
       "intergenic_variant                         4\n",
       "coding_sequence_variant                    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2.consequence.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9bc35074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>gene_id_</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18461</th>\n",
       "      <td>6_99598907_A/C</td>\n",
       "      <td>6:99598907-99598907</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99598907</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>99598906</td>\n",
       "      <td>99598907</td>\n",
       "      <td>6</td>\n",
       "      <td>99606773</td>\n",
       "      <td>99607206</td>\n",
       "      <td>ENSG00000112238</td>\n",
       "      <td>7866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18462</th>\n",
       "      <td>6_99598928_T/C</td>\n",
       "      <td>6:99598928-99598928</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99598928</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>99598927</td>\n",
       "      <td>99598928</td>\n",
       "      <td>6</td>\n",
       "      <td>99606773</td>\n",
       "      <td>99607206</td>\n",
       "      <td>ENSG00000112238</td>\n",
       "      <td>7845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19930</th>\n",
       "      <td>8_11573132_C/T</td>\n",
       "      <td>8:11573132-11573132</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>11573132</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>11573131</td>\n",
       "      <td>11573132</td>\n",
       "      <td>8</td>\n",
       "      <td>11576256</td>\n",
       "      <td>11577317</td>\n",
       "      <td>ENSG00000170983</td>\n",
       "      <td>3124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21511</th>\n",
       "      <td>X_2748343_G/C</td>\n",
       "      <td>X:2748343-2748343</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>2748343</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2748342</td>\n",
       "      <td>2748343</td>\n",
       "      <td>X</td>\n",
       "      <td>2752039</td>\n",
       "      <td>2752335</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>3696</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation             Location Allele         consequence  \\\n",
       "18461      6_99598907_A/C  6:99598907-99598907      -  intergenic_variant   \n",
       "18462      6_99598928_T/C  6:99598928-99598928      -  intergenic_variant   \n",
       "19930      8_11573132_C/T  8:11573132-11573132      -  intergenic_variant   \n",
       "21511       X_2748343_G/C    X:2748343-2748343      -  intergenic_variant   \n",
       "\n",
       "      IMPACT SYMBOL Gene Feature_type Feature BIOTYPE  ...       pos ref alt  \\\n",
       "18461      -      -    -            -       -       -  ...  99598907   A   C   \n",
       "18462      -      -    -            -       -       -  ...  99598928   T   C   \n",
       "19930      -      -    -            -       -       -  ...  11573132   C   T   \n",
       "21511      -      -    -            -       -       -  ...   2748343   G   C   \n",
       "\n",
       "          start       end chrom_    start_      end_         gene_id_ distance  \n",
       "18461  99598906  99598907      6  99606773  99607206  ENSG00000112238     7866  \n",
       "18462  99598927  99598928      6  99606773  99607206  ENSG00000112238     7845  \n",
       "19930  11573131  11573132      8  11576256  11577317  ENSG00000170983     3124  \n",
       "21511   2748342   2748343      X   2752039   2752335  ENSG00000124343     3696  \n",
       "\n",
       "[4 rows x 48 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[V2.consequence == \"intergenic_variant\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0567815a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>gene_id_</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18457</th>\n",
       "      <td>6_99593030_G/T</td>\n",
       "      <td>6:99593030-99593030</td>\n",
       "      <td>-</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99593030</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>99593029</td>\n",
       "      <td>99593030</td>\n",
       "      <td>6</td>\n",
       "      <td>99584879</td>\n",
       "      <td>99589899</td>\n",
       "      <td>ENSG00000279170</td>\n",
       "      <td>3130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18458</th>\n",
       "      <td>6_99593098_A/C</td>\n",
       "      <td>6:99593098-99593098</td>\n",
       "      <td>-</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99593098</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>99593097</td>\n",
       "      <td>99593098</td>\n",
       "      <td>6</td>\n",
       "      <td>99584879</td>\n",
       "      <td>99589899</td>\n",
       "      <td>ENSG00000279170</td>\n",
       "      <td>3198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18459</th>\n",
       "      <td>6_99593111_G/C</td>\n",
       "      <td>6:99593111-99593111</td>\n",
       "      <td>-</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99593111</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>99593110</td>\n",
       "      <td>99593111</td>\n",
       "      <td>6</td>\n",
       "      <td>99584879</td>\n",
       "      <td>99589899</td>\n",
       "      <td>ENSG00000279170</td>\n",
       "      <td>3211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18460</th>\n",
       "      <td>6_99593164_C/T</td>\n",
       "      <td>6:99593164-99593164</td>\n",
       "      <td>-</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>99593164</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>99593163</td>\n",
       "      <td>99593164</td>\n",
       "      <td>6</td>\n",
       "      <td>99584879</td>\n",
       "      <td>99589899</td>\n",
       "      <td>ENSG00000279170</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Uploaded_variation             Location Allele  \\\n",
       "18457      6_99593030_G/T  6:99593030-99593030      -   \n",
       "18458      6_99593098_A/C  6:99593098-99593098      -   \n",
       "18459      6_99593111_G/C  6:99593111-99593111      -   \n",
       "18460      6_99593164_C/T  6:99593164-99593164      -   \n",
       "\n",
       "                     consequence IMPACT SYMBOL Gene Feature_type Feature  \\\n",
       "18457  regulatory_region_variant      -      -    -            -       -   \n",
       "18458  regulatory_region_variant      -      -    -            -       -   \n",
       "18459  regulatory_region_variant      -      -    -            -       -   \n",
       "18460  regulatory_region_variant      -      -    -            -       -   \n",
       "\n",
       "      BIOTYPE  ...       pos ref alt     start       end chrom_    start_  \\\n",
       "18457       -  ...  99593030   G   T  99593029  99593030      6  99584879   \n",
       "18458       -  ...  99593098   A   C  99593097  99593098      6  99584879   \n",
       "18459       -  ...  99593111   G   C  99593110  99593111      6  99584879   \n",
       "18460       -  ...  99593164   C   T  99593163  99593164      6  99584879   \n",
       "\n",
       "           end_         gene_id_ distance  \n",
       "18457  99589899  ENSG00000279170     3130  \n",
       "18458  99589899  ENSG00000279170     3198  \n",
       "18459  99589899  ENSG00000279170     3211  \n",
       "18460  99589899  ENSG00000279170     3264  \n",
       "\n",
       "[4 rows x 48 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2[V2.consequence == \"regulatory_region_variant\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b18bc653",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#Uploaded_variation', 'Location', 'Allele', 'consequence', 'IMPACT',\n",
       "       'SYMBOL', 'Gene', 'Feature_type', 'Feature', 'BIOTYPE', 'EXON',\n",
       "       'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
       "       'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation',\n",
       "       'REF_ALLELE', 'UPLOADED_ALLELE', 'DISTANCE', 'STRAND', 'FLAGS',\n",
       "       'SYMBOL_SOURCE', 'HGNC_ID', 'AF', 'CLIN_SIG', 'SOMATIC', 'PHENO',\n",
       "       'PUBMED', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS',\n",
       "       'MOTIF_SCORE_CHANGE', 'TRANSCRIPTION_FACTORS', 'chrom', 'pos', 'ref',\n",
       "       'alt', 'start', 'end', 'chrom_', 'start_', 'end_', 'gene_id_',\n",
       "       'distance'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "24a56c89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        -\n",
       "1        -\n",
       "2        -\n",
       "3        -\n",
       "4        -\n",
       "        ..\n",
       "23396    -\n",
       "23397    -\n",
       "23398    -\n",
       "23399    -\n",
       "23400    -\n",
       "Name: AF, Length: 23401, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2.AF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b0969db1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AF\n",
       "-    23401\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "V2.AF.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d69eca87",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gpn",
   "language": "python",
   "name": "gpn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
